// payload/dist/admin/components/forms/field-types/Blocks/BlocksDrawer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BlocksDrawer", {
    enumerable: true,
    get: function() {
        return BlocksDrawer;
    }
});
const _modal = require("@faceless-ui/modal");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../../../utilities/getTranslation");
const _Drawer = require("../../../../elements/Drawer");
const _ThumbnailCard = require("../../../../elements/ThumbnailCard");
const _DefaultBlockImage = /*#__PURE__*/ _interop_require_default(require("../../../../graphics/DefaultBlockImage"));
const _BlockSearch = /*#__PURE__*/ _interop_require_default(require("./BlockSearch"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'blocks-drawer';
const getBlockLabel = (block, i18n)=>{
    if (typeof block.labels.singular === 'string') return block.labels.singular.toLowerCase();
    if (typeof block.labels.singular === 'object') {
        return (0, _getTranslation.getTranslation)(block.labels.singular, i18n).toLowerCase();
    }
    return '';
};
const BlocksDrawer = (props)=>{
    const { addRow, addRowIndex, blocks, drawerSlug, labels } = props;
    const [searchTerm, setSearchTerm] = (0, _react.useState)('');
    const [filteredBlocks, setFilteredBlocks] = (0, _react.useState)(blocks);
    const { closeModal, isModalOpen } = (0, _modal.useModal)();
    const { i18n, t } = (0, _reacti18next.useTranslation)('fields');
    (0, _react.useEffect)(()=>{
        if (!isModalOpen) {
            setSearchTerm('');
        }
    }, [
        isModalOpen
    ]);
    (0, _react.useEffect)(()=>{
        const searchTermToUse = searchTerm.toLowerCase();
        const matchingBlocks = blocks.reduce((matchedBlocks, block)=>{
            const blockLabel = getBlockLabel(block, i18n);
            if (blockLabel.includes(searchTermToUse)) matchedBlocks.push(block);
            return matchedBlocks;
        }, []);
        setFilteredBlocks(matchingBlocks);
    }, [
        searchTerm,
        blocks,
        i18n
    ]);
    return /*#__PURE__*/ _react.default.createElement(_Drawer.Drawer, {
        slug: drawerSlug,
        title: t('addLabel', {
            label: (0, _getTranslation.getTranslation)(labels.singular, i18n)
        })
    }, /*#__PURE__*/ _react.default.createElement(_BlockSearch.default, {
        setSearchTerm: setSearchTerm
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__blocks-wrapper`
    }, /*#__PURE__*/ _react.default.createElement("ul", {
        className: `${baseClass}__blocks`
    }, filteredBlocks?.map((block, index)=>{
        const { imageAltText, imageURL, labels: blockLabels, slug } = block;
        return /*#__PURE__*/ _react.default.createElement("li", {
            className: `${baseClass}__block`,
            key: index
        }, /*#__PURE__*/ _react.default.createElement(_ThumbnailCard.ThumbnailCard, {
            alignLabel: "center",
            label: (0, _getTranslation.getTranslation)(blockLabels.singular, i18n),
            onClick: ()=>{
                addRow(addRowIndex, slug);
                closeModal(drawerSlug);
            },
            thumbnail: imageURL ? /*#__PURE__*/ _react.default.createElement("img", {
                alt: imageAltText,
                src: imageURL
            }) : /*#__PURE__*/ _react.default.createElement("div", {
                className: `${baseClass}__default-image`
            }, /*#__PURE__*/ _react.default.createElement(_DefaultBlockImage.default, null))
        }));
    }))));
};
