// payload/dist/fields/hooks/beforeChange/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "beforeChange", {
    enumerable: true,
    get: function() {
        return beforeChange;
    }
});
const _errors = require("../../../errors");
const _deepCopyObject = require("../../../utilities/deepCopyObject");
const _traverseFields = require("./traverseFields");
const beforeChange = async ({ id, collection, context, data: incomingData, doc, docWithLocales, global, operation, req, skipValidation })=>{
    const data = (0, _deepCopyObject.deepCopyObject)(incomingData);
    const mergeLocaleActions = [];
    const errors = [];
    await (0, _traverseFields.traverseFields)({
        id,
        collection,
        context,
        data,
        doc,
        docWithLocales,
        errors,
        fields: collection?.fields || global?.fields,
        global,
        mergeLocaleActions,
        operation,
        path: '',
        req,
        siblingData: data,
        siblingDoc: doc,
        siblingDocWithLocales: docWithLocales,
        skipValidation
    });
    if (errors.length > 0) {
        throw new _errors.ValidationError(errors, req.t);
    }
    mergeLocaleActions.forEach((action)=>action());
    return data;
};
