// payload/dist/admin/components/elements/Pill/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reactrouterdom = require("react-router-dom");
const _useDraggableSortable = require("../DraggableSortable/useDraggableSortable");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'pill';
const DraggablePill = (props)=>{
    const { id, className } = props;
    const { attributes, isDragging, listeners, setNodeRef, transform } = (0, _useDraggableSortable.useDraggableSortable)({
        id
    });
    return /*#__PURE__*/ _react.default.createElement(StaticPill, {
        ...props,
        className: [
            isDragging && `${baseClass}--is-dragging`,
            className
        ].filter(Boolean).join(' '),
        elementProps: {
            ...listeners,
            ...attributes,
            ref: setNodeRef,
            style: {
                transform
            }
        }
    });
};
const StaticPill = (props)=>{
    const { alignIcon = 'right', 'aria-checked': ariaChecked, 'aria-controls': ariaControls, 'aria-expanded': ariaExpanded, 'aria-label': ariaLabel, children, className, draggable, elementProps, icon, onClick, pillStyle = 'light', rounded, to } = props;
    const classes = [
        baseClass,
        `${baseClass}--style-${pillStyle}`,
        className && className,
        to && `${baseClass}--has-link`,
        (to || onClick) && `${baseClass}--has-action`,
        icon && `${baseClass}--has-icon`,
        icon && `${baseClass}--align-icon-${alignIcon}`,
        draggable && `${baseClass}--draggable`,
        rounded && `${baseClass}--rounded`
    ].filter(Boolean).join(' ');
    let Element = 'div';
    if (onClick && !to) Element = 'button';
    if (to) Element = _reactrouterdom.Link;
    return /*#__PURE__*/ _react.default.createElement(Element, {
        ...elementProps,
        "aria-checked": ariaChecked,
        "aria-controls": ariaControls,
        "aria-expanded": ariaExpanded,
        "aria-label": ariaLabel,
        className: classes,
        onClick: onClick,
        to: to || undefined,
        type: Element === 'button' ? 'button' : undefined
    }, icon && alignIcon === 'left' && /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__icon`
    }, icon), /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__label`
    }, children), icon && alignIcon === 'right' && /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__icon`
    }, icon));
};
const Pill = (props)=>{
    const { draggable } = props;
    if (draggable) return /*#__PURE__*/ _react.default.createElement(DraggablePill, props);
    return /*#__PURE__*/ _react.default.createElement(StaticPill, props);
};
const _default = Pill;
