// payload/dist/admin/components/elements/SortColumn/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _getTranslation = require("../../../../utilities/getTranslation");
const _Chevron = /*#__PURE__*/ _interop_require_default(require("../../icons/Chevron"));
const _SearchParams = require("../../utilities/SearchParams");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'sort-column';
const SortColumn = (props)=>{
    const { name, disable = false, label } = props;
    const params = (0, _SearchParams.useSearchParams)();
    const history = (0, _reactrouterdom.useHistory)();
    const { i18n, t } = (0, _reacti18next.useTranslation)('general');
    const { sort } = params;
    const desc = `-${name}`;
    const asc = name;
    const ascClasses = [
        `${baseClass}__asc`
    ];
    if (sort === asc) ascClasses.push(`${baseClass}--active`);
    const descClasses = [
        `${baseClass}__desc`
    ];
    if (sort === desc) descClasses.push(`${baseClass}--active`);
    const setSort = (0, _react.useCallback)((newSort)=>{
        history.push({
            search: _qs.default.stringify({
                ...params,
                sort: newSort
            }, {
                addQueryPrefix: true
            })
        });
    }, [
        params,
        history
    ]);
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__label`
    }, (0, _getTranslation.getTranslation)(label, i18n)), !disable && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__buttons`
    }, /*#__PURE__*/ _react.default.createElement("button", {
        "aria-label": t('sortByLabelDirection', {
            direction: t('ascending'),
            label: (0, _getTranslation.getTranslation)(label, i18n)
        }),
        className: [
            ...ascClasses,
            `${baseClass}__button`
        ].filter(Boolean).join(' '),
        onClick: ()=>setSort(asc),
        type: "button"
    }, /*#__PURE__*/ _react.default.createElement(_Chevron.default, {
        direction: "up"
    })), /*#__PURE__*/ _react.default.createElement("button", {
        "aria-label": t('sortByLabelDirection', {
            direction: t('descending'),
            label: (0, _getTranslation.getTranslation)(label, i18n)
        }),
        className: [
            ...descClasses,
            `${baseClass}__button`
        ].filter(Boolean).join(' '),
        onClick: ()=>setSort(desc),
        type: "button"
    }, /*#__PURE__*/ _react.default.createElement(_Chevron.default, null))));
};
const _default = SortColumn;
