// payload/dist/preferences/operations/update.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _defaultAccess = /*#__PURE__*/ _interop_require_default(require("../../auth/defaultAccess"));
const _executeAccess = /*#__PURE__*/ _interop_require_default(require("../../auth/executeAccess"));
const _UnathorizedError = /*#__PURE__*/ _interop_require_default(require("../../errors/UnathorizedError"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function update(args) {
    const { key, overrideAccess, req: { payload }, req, user, value } = args;
    if (!user) {
        throw new _UnathorizedError.default(req.t);
    }
    const collection = 'payload-preferences';
    const filter = {
        key: {
            equals: key
        },
        'user.relationTo': {
            equals: user.collection
        },
        'user.value': {
            equals: user.id
        }
    };
    const preference = {
        key,
        user: {
            relationTo: user.collection,
            value: user.id
        },
        value
    };
    if (!overrideAccess) {
        await (0, _executeAccess.default)({
            req
        }, _defaultAccess.default);
    }
    try {
        // try/catch because we attempt to update without first reading to check if it exists first to save on db calls
        await payload.db.updateOne({
            collection,
            data: preference,
            req,
            where: filter
        });
    } catch (err) {
        await payload.db.create({
            collection,
            data: preference,
            req
        });
    }
    return preference;
}
const _default = update;
