// payload/dist/admin/components/elements/WhereBuilder/Condition/Number/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _ReactSelect = /*#__PURE__*/ _interop_require_default(require("../../../ReactSelect"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'condition-value-number';
const NumberField = ({ disabled, onChange, operator, value })=>{
    const { t } = (0, _reacti18next.useTranslation)();
    const isMulti = [
        'in',
        'not_in'
    ].includes(operator);
    const [valueToRender, setValueToRender] = _react.default.useState([]);
    const onSelect = _react.default.useCallback((selectedOption)=>{
        let newValue;
        if (!selectedOption) {
            newValue = [];
        } else if (isMulti) {
            if (Array.isArray(selectedOption)) {
                newValue = selectedOption.map((option)=>Number(option.value?.value || option.value));
            } else {
                newValue = [
                    Number(selectedOption.value?.value || selectedOption.value)
                ];
            }
        }
        onChange(newValue);
    }, [
        isMulti,
        onChange
    ]);
    _react.default.useEffect(()=>{
        if (Array.isArray(value)) {
            setValueToRender(value.map((val, index)=>{
                return {
                    id: `${val}${index}`,
                    label: `${val}`,
                    value: {
                        toString: ()=>`${val}${index}`,
                        value: val?.value || val
                    }
                };
            }));
        } else {
            setValueToRender([]);
        }
    }, [
        value
    ]);
    return isMulti ? /*#__PURE__*/ _react.default.createElement(_ReactSelect.default, {
        disabled: disabled,
        isClearable: true,
        isCreatable: true,
        isMulti: isMulti,
        isSortable: true,
        numberOnly: true,
        onChange: onSelect,
        options: [],
        placeholder: t('general:enterAValue'),
        value: valueToRender || []
    }) : /*#__PURE__*/ _react.default.createElement("input", {
        className: baseClass,
        disabled: disabled,
        onChange: (e)=>onChange(e.target.value),
        placeholder: t('general:enterAValue'),
        type: "number",
        value: value
    });
};
const _default = NumberField;
