// payload/dist/admin/components/elements/Nav/context.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    NavContext: function() {
        return NavContext;
    },
    NavProvider: function() {
        return NavProvider;
    },
    useNav: function() {
        return useNav;
    }
});
const _windowinfo = require("@faceless-ui/window-info");
const _bodyscrolllock = require("body-scroll-lock");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reactrouterdom = require("react-router-dom");
const _Preferences = require("../../utilities/Preferences");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const NavContext = /*#__PURE__*/ _react.default.createContext({
    navOpen: true,
    navRef: null,
    setNavOpen: ()=>{}
});
const useNav = ()=>_react.default.useContext(NavContext);
const getNavPreference = async (getPreference)=>{
    const navPrefs = await getPreference('nav');
    const preferredState = navPrefs?.open;
    if (typeof preferredState === 'boolean') {
        return preferredState;
    } else {
        return true;
    }
};
const NavProvider = ({ children })=>{
    const { breakpoints: { l: largeBreak, m: midBreak, s: smallBreak } } = (0, _windowinfo.useWindowInfo)();
    const { getPreference } = (0, _Preferences.usePreferences)();
    const history = (0, _reactrouterdom.useHistory)();
    const navRef = (0, _react.useRef)(null);
    // initialize the nav to be closed
    // this is because getting the preference is async
    // so instead of closing it after the preference is loaded
    // we will open it after the preference is loaded
    const [navOpen, setNavOpen] = _react.default.useState(false);
    // on load check the user's preference and set "initial" state
    (0, _react.useEffect)(()=>{
        if (largeBreak === false) {
            const setNavFromPreferences = async ()=>{
                const preferredState = await getNavPreference(getPreference);
                setNavOpen(preferredState);
            };
            setNavFromPreferences() // eslint-disable-line @typescript-eslint/no-floating-promises
            ;
        }
    }, [
        largeBreak,
        getPreference,
        setNavOpen
    ]);
    // on smaller screens where the nav is a modal
    // close the nav when the user navigates away
    (0, _react.useEffect)(()=>{
        let unlisten;
        if (midBreak) {
            unlisten = history.listen(()=>{
                setNavOpen(false);
            });
        } else if (unlisten) {
            unlisten();
        }
        return ()=>unlisten && unlisten();
    }, [
        history,
        setNavOpen,
        midBreak
    ]);
    // on open and close, lock the body scroll
    // do not do this on desktop, the sidebar is not a modal
    (0, _react.useEffect)(()=>{
        if (navRef.current) {
            if (navOpen && midBreak) {
                (0, _bodyscrolllock.disableBodyScroll)(navRef.current);
            } else {
                (0, _bodyscrolllock.enableBodyScroll)(navRef.current);
            }
        }
    }, [
        navOpen,
        midBreak
    ]);
    // on smaller screens where the nav is a modal
    // close the nav when the user resizes down to mobile
    // the sidebar is a modal on mobile
    (0, _react.useEffect)(()=>{
        if (largeBreak === false || midBreak === false || smallBreak === false) {
            setNavOpen(false);
        }
    }, [
        largeBreak,
        midBreak,
        smallBreak
    ]);
    // when the component unmounts, clear all body scroll locks
    (0, _react.useEffect)(()=>{
        return ()=>{
            (0, _bodyscrolllock.clearAllBodyScrollLocks)();
        };
    }, []);
    return /*#__PURE__*/ _react.default.createElement(NavContext.Provider, {
        value: {
            navOpen,
            navRef,
            setNavOpen
        }
    }, children);
};
