// payload/dist/admin/components/utilities/LoadingOverlay/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    LoadingOverlayProvider: function() {
        return LoadingOverlayProvider;
    },
    default: function() {
        return _default;
    },
    useLoadingOverlay: function() {
        return useLoadingOverlay;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _useDelayedRender = require("../../../hooks/useDelayedRender");
const _Loading = require("../../elements/Loading");
const _reducer = require("./reducer");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const animatedDuration = 250;
const Context = /*#__PURE__*/ (0, _react.createContext)({
    isOnScreen: false,
    toggleLoadingOverlay: undefined
});
const LoadingOverlayProvider = ({ children })=>{
    const { t } = (0, _reacti18next.useTranslation)('general');
    const fallbackText = t('loading');
    const [overlays, dispatchOverlay] = _react.default.useReducer(_reducer.reducer, _reducer.defaultLoadingOverlayState);
    const { isMounted, isUnmounting, triggerDelayedRender } = (0, _useDelayedRender.useDelayedRender)({
        delayBeforeShow: 1000,
        inTimeout: animatedDuration,
        minShowTime: 500,
        outTimeout: animatedDuration,
        show: overlays.isLoading
    });
    const toggleLoadingOverlay = _react.default.useCallback(({ isLoading, key, loadingText = fallbackText, type })=>{
        if (isLoading) {
            triggerDelayedRender();
            dispatchOverlay({
                payload: {
                    key,
                    loadingText,
                    type
                },
                type: 'add'
            });
        } else {
            dispatchOverlay({
                payload: {
                    key,
                    type
                },
                type: 'remove'
            });
        }
    }, [
        triggerDelayedRender,
        fallbackText
    ]);
    return /*#__PURE__*/ _react.default.createElement(Context.Provider, {
        value: {
            isOnScreen: isMounted,
            toggleLoadingOverlay
        }
    }, isMounted && /*#__PURE__*/ _react.default.createElement(_Loading.LoadingOverlay, {
        animationDuration: `${animatedDuration}ms`,
        loadingText: overlays.loadingText || fallbackText,
        overlayType: overlays.overlayType,
        show: !isUnmounting
    }), children);
};
const useLoadingOverlay = ()=>{
    const contextHook = _react.default.useContext(Context);
    if (contextHook === undefined) {
        throw new Error('useLoadingOverlay must be used within a LoadingOverlayProvider');
    }
    return contextHook;
};
const _default = Context;
