// payload/dist/admin/components/utilities/LoadingOverlay/reducer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    defaultLoadingOverlayState: function() {
        return defaultLoadingOverlayState;
    },
    reducer: function() {
        return reducer;
    }
});
const defaultLoadingOverlayState = {
    isLoading: false,
    loaders: [],
    loadingText: '',
    overlayType: null
};
const reducer = (state, action)=>{
    const loadersCopy = [
        ...state.loaders
    ];
    const { key = 'user', loadingText, type = 'fullscreen' } = action.payload;
    if (action.type === 'add') {
        loadersCopy.push({
            key,
            loadingText,
            type
        });
    } else if (action.type === 'remove') {
        const index = loadersCopy.findIndex((item)=>item.key === key && item.type === type);
        loadersCopy.splice(index, 1);
    }
    const nextLoader = loadersCopy?.length > 0 ? loadersCopy[loadersCopy.length - 1] : null;
    return {
        isLoading: Boolean(nextLoader),
        loaders: loadersCopy,
        loadingText: nextLoader?.loadingText || state?.loadingText,
        overlayType: nextLoader?.type || state?.overlayType
    };
};
