// payload/dist/collections/config/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    authDefaults: function() {
        return authDefaults;
    },
    defaults: function() {
        return defaults;
    }
});
const _defaultAccess = /*#__PURE__*/ _interop_require_default(require("../../auth/defaultAccess"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const defaults = {
    access: {
        create: _defaultAccess.default,
        delete: _defaultAccess.default,
        read: _defaultAccess.default,
        unlock: _defaultAccess.default,
        update: _defaultAccess.default
    },
    admin: {
        components: {},
        enableRichTextLink: true,
        enableRichTextRelationship: true,
        pagination: {
            defaultLimit: 10,
            limits: [
                5,
                10,
                25,
                50,
                100
            ]
        },
        useAsTitle: 'id'
    },
    auth: false,
    custom: {},
    endpoints: [],
    fields: [],
    hooks: {
        afterChange: [],
        afterDelete: [],
        afterForgotPassword: [],
        afterLogin: [],
        afterLogout: [],
        afterMe: [],
        afterOperation: [],
        afterRead: [],
        afterRefresh: [],
        beforeChange: [],
        beforeDelete: [],
        beforeLogin: [],
        beforeOperation: [],
        beforeRead: [],
        beforeValidate: [],
        me: [],
        refresh: []
    },
    timestamps: true,
    upload: false,
    versions: false
};
const authDefaults = {
    cookies: {
        sameSite: 'Lax',
        secure: false
    },
    forgotPassword: {},
    lockTime: 600000,
    maxLoginAttempts: 5,
    tokenExpiration: 7200,
    verify: false
};
