// payload/dist/admin/components/templates/Default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _Hamburger = require("../../elements/Hamburger");
const _Header = require("../../elements/Header");
const _Nav = require("../../elements/Nav");
const _NavToggler = require("../../elements/Nav/NavToggler");
const _context = require("../../elements/Nav/context");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
const _RenderCustomComponent = /*#__PURE__*/ _interop_require_default(require("../../utilities/RenderCustomComponent"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'template-default';
const Default = ({ children, className })=>{
    const { admin: { components: { Nav: CustomNav } = {
        Nav: undefined
    } } = {} } = (0, _Config.useConfig)();
    const { t } = (0, _reacti18next.useTranslation)('general');
    const { navOpen } = (0, _context.useNav)();
    return /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: `${t('dashboard')} Payload`,
        keywords: `${t('dashboard')}, Payload`,
        title: t('dashboard')
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__nav-toggler-wrapper`,
        id: "nav-toggler"
    }, /*#__PURE__*/ _react.default.createElement(_NavToggler.NavToggler, {
        className: `${baseClass}__nav-toggler`
    }, /*#__PURE__*/ _react.default.createElement(_Hamburger.Hamburger, {
        closeIcon: "collapse",
        isActive: navOpen
    }))), /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            className,
            navOpen && `${baseClass}--nav-open`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_RenderCustomComponent.default, {
        CustomComponent: CustomNav,
        DefaultComponent: _Nav.Nav
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrap`
    }, /*#__PURE__*/ _react.default.createElement(_Header.AppHeader, null), children)));
};
const _default = Default;
