// payload/dist/versions/drafts/replaceWithDraftIfAvailable.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _auth = require("../../auth");
const _combineQueries = require("../../database/combineQueries");
const _types = require("../../types");
const _sanitizeInternalFields = /*#__PURE__*/ _interop_require_default(require("../../utilities/sanitizeInternalFields"));
const _appendVersionToQueryKey = require("./appendVersionToQueryKey");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const replaceWithDraftIfAvailable = async ({ accessResult, doc, entity, entityType, req })=>{
    const { locale } = req;
    const queryToBuild = {
        and: [
            {
                'version._status': {
                    equals: 'draft'
                }
            }
        ]
    };
    if (entityType === 'collection') {
        queryToBuild.and.push({
            parent: {
                equals: doc.id
            }
        });
    }
    if ((0, _types.docHasTimestamps)(doc)) {
        queryToBuild.and.push({
            updatedAt: {
                greater_than: doc.updatedAt
            }
        });
    }
    let versionAccessResult;
    if ((0, _auth.hasWhereAccessResult)(accessResult)) {
        versionAccessResult = (0, _appendVersionToQueryKey.appendVersionToQueryKey)(accessResult);
    }
    const findVersionsArgs = {
        collection: entity.slug,
        global: entity.slug,
        limit: 1,
        locale,
        req,
        sort: '-updatedAt',
        where: (0, _combineQueries.combineQueries)(queryToBuild, versionAccessResult)
    };
    let versionDocs;
    if (entityType === 'global') {
        versionDocs = (await req.payload.db.findGlobalVersions(findVersionsArgs)).docs;
    } else {
        versionDocs = (await req.payload.db.findVersions(findVersionsArgs)).docs;
    }
    let draft = versionDocs[0];
    if (!draft) {
        return doc;
    }
    draft = JSON.parse(JSON.stringify(draft));
    draft = (0, _sanitizeInternalFields.default)(draft);
    // Patch globalType onto version doc
    if (entityType === 'global' && 'globalType' in doc) {
        draft.version.globalType = doc.globalType;
    }
    // Disregard all other draft content at this point,
    // Only interested in the version itself.
    // Operations will handle firing hooks, etc.
    return {
        id: doc.id,
        ...draft.version,
        createdAt: draft.createdAt,
        updatedAt: draft.updatedAt
    };
};
const _default = replaceWithDraftIfAvailable;
