// payload/dist/admin/components/views/ResetPassword/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _reacttoastify = require("react-toastify");
const _Button = /*#__PURE__*/ _interop_require_default(require("../../elements/Button"));
const _Form = /*#__PURE__*/ _interop_require_default(require("../../forms/Form"));
const _Submit = /*#__PURE__*/ _interop_require_default(require("../../forms/Submit"));
const _ConfirmPassword = /*#__PURE__*/ _interop_require_default(require("../../forms/field-types/ConfirmPassword"));
const _HiddenInput = /*#__PURE__*/ _interop_require_default(require("../../forms/field-types/HiddenInput"));
const _Password = /*#__PURE__*/ _interop_require_default(require("../../forms/field-types/Password"));
const _Minimal = /*#__PURE__*/ _interop_require_default(require("../../templates/Minimal"));
const _Auth = require("../../utilities/Auth");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'reset-password';
const ResetPassword = ()=>{
    const config = (0, _Config.useConfig)();
    const { admin: { logoutRoute, user: userSlug }, routes: { admin, api }, serverURL } = config;
    const { token } = (0, _reactrouterdom.useParams)();
    const history = (0, _reactrouterdom.useHistory)();
    const { fetchFullUser, user } = (0, _Auth.useAuth)();
    const { t } = (0, _reacti18next.useTranslation)('authentication');
    const onSuccess = async (data)=>{
        if (data.token) {
            await fetchFullUser();
            history.push(`${admin}`);
        } else {
            history.push(`${admin}/login`);
            _reacttoastify.toast.success(t('general:updatedSuccessfully'), {
                autoClose: 3000
            });
        }
    };
    if (user) {
        return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
            className: baseClass
        }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
            description: t('resetPassword'),
            keywords: t('resetPassword'),
            title: t('resetPassword')
        }), /*#__PURE__*/ _react.default.createElement("div", {
            className: `${baseClass}__wrap`
        }, /*#__PURE__*/ _react.default.createElement("h1", null, t('alreadyLoggedIn')), /*#__PURE__*/ _react.default.createElement("p", null, /*#__PURE__*/ _react.default.createElement(_reacti18next.Trans, {
            i18nKey: "loginWithAnotherUser",
            t: t
        }, /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
            to: `${admin}${logoutRoute}`
        }, "log out"))), /*#__PURE__*/ _react.default.createElement("br", null), /*#__PURE__*/ _react.default.createElement(_Button.default, {
            buttonStyle: "secondary",
            el: "link",
            to: admin
        }, t('general:backToDashboard'))));
    }
    return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrap`
    }, /*#__PURE__*/ _react.default.createElement("h1", null, t('resetPassword')), /*#__PURE__*/ _react.default.createElement(_Form.default, {
        action: `${serverURL}${api}/${userSlug}/reset-password`,
        method: "post",
        onSuccess: onSuccess,
        redirect: admin
    }, /*#__PURE__*/ _react.default.createElement(_Password.default, {
        autoComplete: "off",
        label: t('newPassword'),
        name: "password",
        required: true
    }), /*#__PURE__*/ _react.default.createElement(_ConfirmPassword.default, null), /*#__PURE__*/ _react.default.createElement(_HiddenInput.default, {
        name: "token",
        value: token
    }), /*#__PURE__*/ _react.default.createElement(_Submit.default, null, t('resetPassword')))));
};
const _default = ResetPassword;
