// @faceless-ui/modal/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useModal = exports.ModalToggler = exports.ModalProvider = exports.ModalContext = exports.ModalContainer = exports.Modal = exports.asModal = void 0;
var asModal_1 = require("./asModal");
Object.defineProperty(exports, "asModal", { enumerable: true, get: function () { return __importDefault(asModal_1).default; } });
var Modal_1 = require("./Modal");
Object.defineProperty(exports, "Modal", { enumerable: true, get: function () { return __importDefault(Modal_1).default; } });
var ModalContainer_1 = require("./ModalContainer");
Object.defineProperty(exports, "ModalContainer", { enumerable: true, get: function () { return __importDefault(ModalContainer_1).default; } });
var context_1 = require("./ModalProvider/context");
Object.defineProperty(exports, "ModalContext", { enumerable: true, get: function () { return __importDefault(context_1).default; } });
var ModalProvider_1 = require("./ModalProvider");
Object.defineProperty(exports, "ModalProvider", { enumerable: true, get: function () { return __importDefault(ModalProvider_1).default; } });
var ModalToggler_1 = require("./ModalToggler");
Object.defineProperty(exports, "ModalToggler", { enumerable: true, get: function () { return __importDefault(ModalToggler_1).default; } });
var useModal_1 = require("./useModal");
Object.defineProperty(exports, "useModal", { enumerable: true, get: function () { return __importDefault(useModal_1).default; } });
