// payload/dist/admin/components/views/LivePreview/Toolbar/SizeInput/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PreviewFrameSizeInput", {
    enumerable: true,
    get: function() {
        return PreviewFrameSizeInput;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _context = require("../../Context/context");
require("./index.scss");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'toolbar-input';
const PreviewFrameSizeInput = (props)=>{
    const { axis } = props;
    const { breakpoint, measuredDeviceSize, setBreakpoint, setSize, size, zoom } = (0, _context.useLivePreviewContext)();
    const [internalState, setInternalState] = _react.default.useState((axis === 'x' ? measuredDeviceSize?.width : measuredDeviceSize?.height) || 0);
    // when the input is changed manually, we need to set the breakpoint as `custom`
    // this will then allow us to set an explicit width and height
    const handleChange = (0, _react.useCallback)((e)=>{
        let newValue = Number(e.target.value);
        if (newValue < 0) newValue = 0;
        setInternalState(newValue);
        setBreakpoint('custom');
        // be sure to set _both_ axis values to so that the other axis doesn't fallback to 0 on initial change
        // this is because the `responsive` size is '100%' in CSS, and `0` in initial state
        setSize({
            type: 'reset',
            value: {
                height: axis === 'y' ? newValue : Number(measuredDeviceSize?.height.toFixed(0)) * zoom,
                width: axis === 'x' ? newValue : Number(measuredDeviceSize?.width.toFixed(0)) * zoom
            }
        });
    }, [
        axis,
        setBreakpoint,
        measuredDeviceSize,
        setSize,
        zoom
    ]);
    // if the breakpoint is `responsive` then the device's div will have `100%` width and height
    // so we need to take the measurements provided by `actualDeviceSize` and sync internal state
    (0, _react.useEffect)(()=>{
        if (breakpoint === 'responsive' && measuredDeviceSize) {
            if (axis === 'x') setInternalState(Number(measuredDeviceSize.width.toFixed(0)) * zoom);
            else setInternalState(Number(measuredDeviceSize.height.toFixed(0)) * zoom);
        }
        if (breakpoint !== 'responsive' && size) {
            setInternalState(axis === 'x' ? size.width : size.height);
        }
    }, [
        breakpoint,
        axis,
        measuredDeviceSize,
        size,
        zoom
    ]);
    return /*#__PURE__*/ _react.default.createElement("input", {
        className: baseClass,
        min: 0,
        name: axis === 'x' ? 'live-preview-width' : 'live-preview-height',
        onChange: handleChange,
        step: 1,
        type: "number",
        value: internalState || 0
    });
};
