// payload/dist/bin/index.js
/* eslint-disable @typescript-eslint/no-floating-promises */ /* eslint-disable @typescript-eslint/no-var-requires */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
const _register = /*#__PURE__*/ _interop_require_default(require("@swc/register"));
const _dotenv = /*#__PURE__*/ _interop_require_default(require("dotenv"));
const _findup = /*#__PURE__*/ _interop_require_default(require("find-up"));
const _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
const _gettsconfig = require("get-tsconfig");
const _minimist = /*#__PURE__*/ _interop_require_default(require("minimist"));
const _path = /*#__PURE__*/ _interop_require_default(require("path"));
const _generateGraphQLSchema = require("./generateGraphQLSchema");
const _generateTypes = require("./generateTypes");
const _migrate = require("./migrate");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
loadEnv();
const tsConfig = (0, _gettsconfig.getTsconfig)();
const swcOptions = {
    ignore: [
        /.*[\\/]node_modules[\\/].*/
    ],
    jsc: {
        baseUrl: _path.default.resolve(),
        parser: {
            syntax: 'typescript',
            tsx: true
        },
        paths: undefined
    },
    module: {
        type: 'commonjs'
    },
    sourceMaps: 'inline'
};
if (tsConfig?.config?.compilerOptions?.paths) {
    swcOptions.jsc.paths = tsConfig.config.compilerOptions.paths;
    if (tsConfig?.config?.compilerOptions?.baseUrl) {
        swcOptions.jsc.baseUrl = _path.default.resolve(tsConfig.config.compilerOptions.baseUrl);
    }
}
// Allow disabling SWC for debugging
if (process.env.DISABLE_SWC !== 'true') {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error - bad @swc/register types
    (0, _register.default)(swcOptions);
}
const { build } = require('./build');
const args = (0, _minimist.default)(process.argv.slice(2));
const scriptIndex = args._.findIndex((x)=>x === 'build');
const script = scriptIndex === -1 ? args._[0] : args._[scriptIndex];
if (script) {
    if (script.startsWith('migrate')) {
        (0, _migrate.migrate)(args).then(()=>process.exit(0));
    } else {
        switch(script.toLowerCase()){
            case 'build':
                {
                    build();
                    break;
                }
            case 'generate:types':
                {
                    (0, _generateTypes.generateTypes)();
                    break;
                }
            case 'generate:graphqlschema':
                {
                    (0, _generateGraphQLSchema.generateGraphQLSchema)();
                    break;
                }
            default:
                console.log(`Unknown script "${script}".`);
                break;
        }
    }
} else {
    console.error('No payload script specified. Did you mean to run `payload migrate`?');
}
/**
 * Try to find user's .env and load it
 */ function loadEnv() {
    const envPath = _findup.default.sync('.env');
    if (envPath) {
        _dotenv.default.config({
            path: envPath
        });
    } else {
        const cwdPath = _path.default.resolve(process.cwd(), '.env');
        if (_fs.default.existsSync(cwdPath)) {
            _dotenv.default.config({
                path: cwdPath
            });
        }
    }
}
