// payload/dist/bin/migrate.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "migrate", {
    enumerable: true,
    get: function() {
        return migrate;
    }
});
const _minimist = /*#__PURE__*/ _interop_require_default(require("minimist"));
const _ = /*#__PURE__*/ _interop_require_default(require(".."));
const _logger = require("../utilities/logger");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/**
 * The default logger's options did not allow for forcing sync logging
 * Using these options, to force both pretty print and sync logging
 */ const prettySyncLogger = {
    loggerDestination: _logger.prettySyncLoggerDestination,
    loggerOptions: {}
};
const availableCommands = [
    'migrate',
    'migrate:create',
    'migrate:down',
    'migrate:refresh',
    'migrate:reset',
    'migrate:status',
    'migration:fresh'
];
const availableCommandsMsg = `Available commands: ${availableCommands.join(', ')}`;
const migrate = async (parsedArgs)=>{
    const { _: args, file, forceAcceptWarning, help } = parsedArgs;
    if (help) {
        // eslint-disable-next-line no-console
        console.log(`\n\n${availableCommandsMsg}\n`) // Avoid having to init payload to get the logger
        ;
        process.exit(0);
    }
    process.env.PAYLOAD_MIGRATING = 'true';
    // Barebones instance to access database adapter
    await _.default.init({
        disableOnInit: true,
        local: true,
        secret: process.env.PAYLOAD_SECRET || '--unused--',
        ...prettySyncLogger
    });
    const adapter = _.default.db;
    if (!adapter) {
        throw new Error('No database adapter found');
    }
    if (!args.length) {
        _.default.logger.error({
            msg: `No migration command provided. ${availableCommandsMsg}`
        });
        process.exit(1);
    }
    switch(args[0]){
        case 'migrate':
            await adapter.migrate();
            break;
        case 'migrate:status':
            await adapter.migrateStatus();
            break;
        case 'migrate:down':
            await adapter.migrateDown();
            break;
        case 'migrate:refresh':
            await adapter.migrateRefresh();
            break;
        case 'migrate:reset':
            await adapter.migrateReset();
            break;
        case 'migrate:fresh':
            await adapter.migrateFresh({
                forceAcceptWarning
            });
            break;
        case 'migrate:create':
            try {
                await adapter.createMigration({
                    file,
                    forceAcceptWarning,
                    migrationName: args[1],
                    payload: _.default
                });
            } catch (err) {
                throw new Error(`Error creating migration: ${err.message}`);
            }
            break;
        default:
            _.default.logger.error({
                msg: `Unknown migration command: ${args[0]}. ${availableCommandsMsg}`
            });
            process.exit(1);
    }
    _.default.logger.info('Done.');
};
// When launched directly call migrate
if (module.id === require.main.id) {
    const args = (0, _minimist.default)(process.argv.slice(2));
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    migrate(args).then(()=>{
        process.exit(0);
    });
}
