// payload/dist/collections/operations/local/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _local = /*#__PURE__*/ _interop_require_default(require("../../../auth/operations/local"));
const _count = /*#__PURE__*/ _interop_require_default(require("./count"));
const _create = /*#__PURE__*/ _interop_require_default(require("./create"));
const _delete = /*#__PURE__*/ _interop_require_default(require("./delete"));
const _find = /*#__PURE__*/ _interop_require_default(require("./find"));
const _findByID = /*#__PURE__*/ _interop_require_default(require("./findByID"));
const _findVersionByID = /*#__PURE__*/ _interop_require_default(require("./findVersionByID"));
const _findVersions = /*#__PURE__*/ _interop_require_default(require("./findVersions"));
const _restoreVersion = /*#__PURE__*/ _interop_require_default(require("./restoreVersion"));
const _update = /*#__PURE__*/ _interop_require_default(require("./update"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const _default = {
    auth: _local.default,
    count: _count.default,
    create: _create.default,
    deleteLocal: _delete.default,
    find: _find.default,
    findByID: _findByID.default,
    findVersionByID: _findVersionByID.default,
    findVersions: _findVersions.default,
    restoreVersion: _restoreVersion.default,
    update: _update.default
};
