// payload/dist/graphql/schema/buildPaginatedListType.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _graphql = require("graphql");
const buildPaginatedListType = (name, docType)=>new _graphql.GraphQLObjectType({
        name,
        fields: {
            docs: {
                type: new _graphql.GraphQLList(docType)
            },
            hasNextPage: {
                type: _graphql.GraphQLBoolean
            },
            hasPrevPage: {
                type: _graphql.GraphQLBoolean
            },
            limit: {
                type: _graphql.GraphQLInt
            },
            nextPage: {
                type: _graphql.GraphQLInt
            },
            offset: {
                type: _graphql.GraphQLInt
            },
            page: {
                type: _graphql.GraphQLInt
            },
            pagingCounter: {
                type: _graphql.GraphQLInt
            },
            prevPage: {
                type: _graphql.GraphQLInt
            },
            totalDocs: {
                type: _graphql.GraphQLInt
            },
            totalPages: {
                type: _graphql.GraphQLInt
            }
        }
    });
const _default = buildPaginatedListType;
