// payload/dist/admin/components/views/Versions/Default.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DefaultVersionsView", {
    enumerable: true,
    get: function() {
        return DefaultVersionsView;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _Gutter = require("../../elements/Gutter");
const _Loading = require("../../elements/Loading");
const _Paginator = /*#__PURE__*/ _interop_require_default(require("../../elements/Paginator"));
const _PerPage = /*#__PURE__*/ _interop_require_default(require("../../elements/PerPage"));
const _Table = require("../../elements/Table");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
const _SearchParams = require("../../utilities/SearchParams");
const _SetStepNav = require("../collections/Edit/SetStepNav");
const _columns = require("./columns");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'versions';
const DefaultVersionsView = (props)=>{
    const { id, collection, data, entityLabel, global, isLoadingVersions, latestDraftVersion, latestPublishedVersion, versionsData } = props;
    const { t } = (0, _reacti18next.useTranslation)('version');
    const { limit } = (0, _SearchParams.useSearchParams)();
    const useAsTitle = collection?.admin?.useAsTitle || 'id';
    let metaDesc;
    let metaTitle;
    if (collection) {
        metaTitle = `${t('versions')} - ${data[useAsTitle]} - ${entityLabel}`;
        metaDesc = t('viewingVersions', {
            documentTitle: data[useAsTitle],
            entityLabel
        });
    }
    if (global) {
        metaTitle = `${t('versions')} - ${entityLabel}`;
        metaDesc = t('viewingVersionsGlobal', {
            entityLabel
        });
    }
    const versionCount = versionsData?.totalDocs || 0;
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_SetStepNav.SetStepNav, {
        collection: collection,
        global: global,
        id: id,
        isEditing: true,
        view: t('versions')
    }), /*#__PURE__*/ _react.default.createElement(_Loading.LoadingOverlayToggle, {
        name: "versions",
        show: isLoadingVersions
    }), /*#__PURE__*/ _react.default.createElement("main", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: metaDesc,
        title: metaTitle
    }), /*#__PURE__*/ _react.default.createElement(_Gutter.Gutter, {
        className: `${baseClass}__wrap`
    }, versionCount === 0 && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__no-versions`
    }, t('noFurtherVersionsFound')), versionCount > 0 && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Table.Table, {
        columns: (0, _columns.buildVersionColumns)(collection, global, t, latestDraftVersion, latestPublishedVersion),
        data: versionsData?.docs
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__page-controls`
    }, /*#__PURE__*/ _react.default.createElement(_Paginator.default, {
        hasNextPage: versionsData.hasNextPage,
        hasPrevPage: versionsData.hasPrevPage,
        limit: versionsData.limit,
        nextPage: versionsData.nextPage,
        numberOfNeighbors: 1,
        page: versionsData.page,
        prevPage: versionsData.prevPage,
        totalPages: versionsData.totalPages
    }), versionsData?.totalDocs > 0 && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__page-info`
    }, versionsData.page * versionsData.limit - (versionsData.limit - 1), "-", versionsData.totalPages > 1 && versionsData.totalPages !== versionsData.page ? versionsData.limit * versionsData.page : versionsData.totalDocs, ' ', t('of'), " ", versionsData.totalDocs), /*#__PURE__*/ _react.default.createElement(_PerPage.default, {
        limit: limit ? Number(limit) : 10,
        limits: collection?.admin?.pagination?.limits
    })))))));
};
