// payload/dist/globals/graphql/resolvers/findVersions.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findVersionsResolver;
    }
});
const _isolateObjectProperty = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateObjectProperty"));
const _findVersions = /*#__PURE__*/ _interop_require_default(require("../../operations/findVersions"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function findVersionsResolver(globalConfig) {
    return async function resolver(_, args, context) {
        const options = {
            depth: 0,
            globalConfig,
            limit: args.limit,
            page: args.page,
            req: (0, _isolateObjectProperty.default)(context.req, 'transactionID'),
            sort: args.sort,
            where: args.where
        };
        const result = await (0, _findVersions.default)(options);
        return result;
    };
}
