// payload/dist/admin/components/elements/Logout/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _LogOut = /*#__PURE__*/ _interop_require_default(require("../../icons/LogOut"));
const _Config = require("../../utilities/Config");
const _RenderCustomComponent = /*#__PURE__*/ _interop_require_default(require("../../utilities/RenderCustomComponent"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'nav';
const DefaultLogout = ({ tabIndex })=>{
    const { t } = (0, _reacti18next.useTranslation)('authentication');
    const config = (0, _Config.useConfig)();
    const { admin: { logoutRoute }, routes: { admin } } = config;
    return /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
        "aria-label": t('logOut'),
        className: `${baseClass}__log-out`,
        tabIndex: tabIndex,
        to: `${admin}${logoutRoute}`
    }, /*#__PURE__*/ _react.default.createElement(_LogOut.default, null));
};
const Logout = ({ tabIndex = 0 })=>{
    const { admin: { components: { logout: { Button: CustomLogout } = {
        Button: undefined
    } } = {} } = {} } = (0, _Config.useConfig)();
    return /*#__PURE__*/ _react.default.createElement(_RenderCustomComponent.default, {
        CustomComponent: CustomLogout,
        DefaultComponent: DefaultLogout,
        componentProps: {
            tabIndex
        }
    });
};
const _default = Logout;
