// payload/dist/admin/components/views/Routes/collections.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "collectionRoutes", {
    enumerable: true,
    get: function() {
        return collectionRoutes;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reactrouterdom = require("react-router-dom");
const _DocumentInfo = require("../../utilities/DocumentInfo");
const _List = /*#__PURE__*/ _interop_require_default(require("../collections/List"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
// @ts-expect-error Just TypeScript being broken // TODO: Open TypeScript issue
const Edit = /*#__PURE__*/ (0, _react.lazy)(()=>Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("../collections/Edit"))));
// @ts-expect-error Just TypeScript being broken // TODO: Open TypeScript issue
const Unauthorized = /*#__PURE__*/ (0, _react.lazy)(()=>Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("../Unauthorized"))));
const collectionRoutes = (props)=>{
    const { collections, match, permissions, user } = props;
    // Note: `Route` must be directly nested within `Switch` for `useRouteMatch` to work
    // This means that we cannot use `Fragment` here with a simple map function to return an array of routes
    // Instead, we need to use `reduce` to return an array of routes directly within `Switch`
    return collections?.filter(({ admin: { hidden } })=>!(typeof hidden === 'function' ? hidden({
            user
        }) : hidden)).reduce((acc, collection)=>{
        // Default routes
        const routesToReturn = [
            ...acc,
            /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Route, {
                exact: true,
                key: `${collection.slug}-list`,
                path: `${match.url}/collections/${collection.slug}`
            }, permissions?.collections?.[collection.slug]?.read?.permission ? /*#__PURE__*/ _react.default.createElement(_List.default, {
                collection: collection
            }) : /*#__PURE__*/ _react.default.createElement(Unauthorized, null)),
            /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Route, {
                exact: true,
                key: `${collection.slug}-create`,
                path: `${match.url}/collections/${collection.slug}/create`
            }, permissions?.collections?.[collection.slug]?.create?.permission ? /*#__PURE__*/ _react.default.createElement(_DocumentInfo.DocumentInfoProvider, {
                collection: collection,
                idFromParams: true
            }, /*#__PURE__*/ _react.default.createElement(Edit, {
                collection: collection
            })) : /*#__PURE__*/ _react.default.createElement(Unauthorized, null)),
            /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Route, {
                key: `${collection.slug}-edit`,
                path: `${match.url}/collections/${collection.slug}/:id`
            }, permissions?.collections?.[collection.slug]?.read?.permission ? /*#__PURE__*/ _react.default.createElement(_DocumentInfo.DocumentInfoProvider, {
                collection: collection,
                idFromParams: true
            }, /*#__PURE__*/ _react.default.createElement(Edit, {
                collection: collection,
                isEditing: true
            })) : /*#__PURE__*/ _react.default.createElement(Unauthorized, null))
        ];
        return routesToReturn;
    }, []);
};
