// payload/dist/admin/components/views/Routes/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "customRoutes", {
    enumerable: true,
    get: function() {
        return customRoutes;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reactrouterdom = require("react-router-dom");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const internalViews = [
    'Account',
    'Dashboard'
];
const customRoutes = (props)=>{
    const { canAccessAdmin, config, match, user } = props;
    const { admin: { components: { views: customViewConfigs } = {} } } = config;
    const customViews = Object.entries(customViewConfigs || {}).filter(([viewKey, view])=>{
        // Remove internal views from the list of custom views
        // This way we can easily iterate over the remaining views
        return Boolean(!internalViews.includes(viewKey) && typeof view !== 'function' && typeof view === 'object');
    })?.map(([, view])=>view);
    if (Array.isArray(customViews)) {
        return customViews.map((CustomView)=>{
            // You are responsible for ensuring that your own custom route is secure
            // i.e. return `Unauthorized` in your own component if the user does not have permission
            if (typeof CustomView === 'function') {
                return /*#__PURE__*/ _react.default.createElement(CustomView, {
                    user: user
                });
            }
            const { Component, exact, path, sensitive, strict } = CustomView;
            return /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Route, {
                exact: exact,
                key: `${match.url}${path}`,
                path: `${match.url}${path}`,
                sensitive: sensitive,
                strict: strict
            }, /*#__PURE__*/ _react.default.createElement(Component, {
                canAccessAdmin: canAccessAdmin,
                user: user
            }));
        });
    }
    return null;
};
