// payload/dist/admin/components/elements/PerPage/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _defaults = require("../../../../collections/config/defaults");
const _Chevron = /*#__PURE__*/ _interop_require_default(require("../../icons/Chevron"));
const _SearchParams = require("../../utilities/SearchParams");
const _Popup = /*#__PURE__*/ _interop_require_default(require("../Popup"));
const _PopupButtonList = /*#__PURE__*/ _interop_require_wildcard(require("../Popup/PopupButtonList"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'per-page';
const defaultLimits = _defaults.defaults.admin.pagination.limits;
const PerPage = ({ handleChange, limit, limits = defaultLimits, modifySearchParams = true, resetPage = false })=>{
    const params = (0, _SearchParams.useSearchParams)();
    const history = (0, _reactrouterdom.useHistory)();
    const { t } = (0, _reacti18next.useTranslation)('general');
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement(_Popup.default, {
        button: /*#__PURE__*/ _react.default.createElement("div", {
            className: `${baseClass}__base-button`
        }, /*#__PURE__*/ _react.default.createElement("span", null, t('perPage', {
            limit
        })), " ", /*#__PURE__*/ _react.default.createElement(_Chevron.default, {
            className: `${baseClass}__icon`
        })),
        horizontalAlign: "right",
        render: ({ close })=>/*#__PURE__*/ _react.default.createElement(_PopupButtonList.ButtonGroup, null, limits.map((limitNumber, i)=>/*#__PURE__*/ _react.default.createElement(_PopupButtonList.Button, {
                    className: [
                        `${baseClass}__button`,
                        limitNumber === Number(limit) && `${baseClass}__button-active`
                    ].filter(Boolean).join(' '),
                    key: i,
                    onClick: ()=>{
                        close();
                        if (handleChange) handleChange(limitNumber);
                        if (modifySearchParams) {
                            history.replace({
                                search: _qs.default.stringify({
                                    ...params,
                                    limit: limitNumber,
                                    page: resetPage ? 1 : params.page
                                }, {
                                    addQueryPrefix: true
                                })
                            });
                        }
                    }
                }, limitNumber === Number(limit) && /*#__PURE__*/ _react.default.createElement("div", {
                    className: `${baseClass}__chevron`
                }, /*#__PURE__*/ _react.default.createElement(_Chevron.default, {
                    direction: "right"
                })), " ", /*#__PURE__*/ _react.default.createElement("span", null, limitNumber)))),
        size: "small"
    }));
};
const _default = PerPage;
