// payload/dist/admin/components/elements/TableColumns/buildColumns.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _types = require("../../../../fields/config/types");
const _flattenTopLevelFields = /*#__PURE__*/ _interop_require_default(require("../../../../utilities/flattenTopLevelFields"));
const _Cell = /*#__PURE__*/ _interop_require_default(require("../../views/collections/List/Cell"));
const _SelectAll = /*#__PURE__*/ _interop_require_default(require("../../views/collections/List/SelectAll"));
const _SelectRow = /*#__PURE__*/ _interop_require_default(require("../../views/collections/List/SelectRow"));
const _SortColumn = /*#__PURE__*/ _interop_require_default(require("../SortColumn"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const buildColumns = ({ cellProps, collection, columns })=>{
    // sort the fields to the order of activeColumns
    const sortedFields = (0, _flattenTopLevelFields.default)(collection.fields, true).sort((a, b)=>{
        const aIndex = columns.findIndex((column)=>column.accessor === a.name);
        const bIndex = columns.findIndex((column)=>column.accessor === b.name);
        if (aIndex === -1 && bIndex === -1) return 0;
        if (aIndex === -1) return 1;
        if (bIndex === -1) return -1;
        return aIndex - bIndex;
    });
    const firstActiveColumn = sortedFields.find((field)=>columns.find((column)=>column.accessor === field.name)?.active);
    let colIndex = -1;
    const cols = sortedFields.map((field)=>{
        const isActive = columns.find((column)=>column.accessor === field.name)?.active || false;
        const isFirstActive = firstActiveColumn?.name === field.name;
        if (isActive) {
            colIndex += 1;
        }
        const props = cellProps?.[colIndex] || {};
        const fieldAffectsDataSubFields = field && field.type && (field.type === 'array' || field.type === 'group' || field.type === 'blocks');
        const disableListFilter = field.admin && 'disableListFilter' in field.admin ? field.admin.disableListFilter : false;
        return {
            name: field.name,
            accessor: field.name,
            active: isActive,
            admin: {
                disableListColumn: field.admin?.disableListColumn,
                disableListFilter
            },
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    disable: fieldAffectsDataSubFields || (0, _types.fieldIsPresentationalOnly)(field) || undefined,
                    label: field.label || field.name,
                    name: field.name
                }),
                renderCell: (rowData, cellData)=>{
                    return /*#__PURE__*/ _react.default.createElement(_Cell.default, {
                        cellData: cellData,
                        colIndex: colIndex,
                        collection: collection,
                        field: field,
                        key: JSON.stringify(cellData),
                        link: isFirstActive,
                        rowData: rowData,
                        ...props
                    });
                }
            },
            label: field.label
        };
    });
    if (cellProps?.[0]?.link !== false) {
        cols.unshift({
            name: '',
            accessor: '_select',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SelectAll.default, null),
                renderCell: (rowData)=>/*#__PURE__*/ _react.default.createElement(_SelectRow.default, {
                        id: rowData.id
                    })
            },
            label: null
        });
    }
    return cols;
};
const _default = buildColumns;
