// payload/components/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Meta: function() {
        return _Meta.default;
    },
    buildStateFromSchema: function() {
        return _buildStateFromSchema.default;
    },
    useAuth: function() {
        return _Auth.useAuth;
    },
    useCollapsible: function() {
        return _provider.useCollapsible;
    },
    useConfig: function() {
        return _Config.useConfig;
    },
    useDocumentInfo: function() {
        return _DocumentInfo.useDocumentInfo;
    },
    useEditDepth: function() {
        return _EditDepth.useEditDepth;
    },
    useLocale: function() {
        return _Locale.useLocale;
    },
    useTheme: function() {
        return _Theme.useTheme;
    },
    withMergedProps: function() {
        return _WithMergedProps.withMergedProps;
    }
});
const _provider = require("../dist/admin/components/elements/Collapsible/provider");
const _buildStateFromSchema = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Form/buildStateFromSchema"));
const _Auth = require("../dist/admin/components/utilities/Auth");
const _Config = require("../dist/admin/components/utilities/Config");
const _DocumentInfo = require("../dist/admin/components/utilities/DocumentInfo");
const _EditDepth = require("../dist/admin/components/utilities/EditDepth");
const _Locale = require("../dist/admin/components/utilities/Locale");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/utilities/Meta"));
const _Theme = require("../dist/admin/components/utilities/Theme");
const _WithMergedProps = require("../dist/admin/components/utilities/WithMergedProps");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
