// payload/dist/collections/requestHandlers/deleteByID.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return deleteByIDHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _errors = require("../../errors");
const _sanitizeCollectionID = require("../../utilities/sanitizeCollectionID");
const _deleteByID = /*#__PURE__*/ _interop_require_default(require("../operations/deleteByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function deleteByIDHandler(req, res, next) {
    const id = (0, _sanitizeCollectionID.sanitizeCollectionID)({
        id: req.params.id,
        collectionSlug: req.collection.config.slug,
        payload: req.payload
    });
    try {
        const doc = await (0, _deleteByID.default)({
            id,
            collection: req.collection,
            depth: parseInt(String(req.query.depth), 10),
            req
        });
        if (!doc) {
            res.status(_httpstatus.default.NOT_FOUND).json(new _errors.NotFound(req.t));
        }
        res.status(_httpstatus.default.OK).send(doc);
    } catch (error) {
        next(error);
    }
}
