// payload/dist/collections/requestHandlers/findVersionByID.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findVersionByIDHandler;
    }
});
const _sanitizeCollectionID = require("../../utilities/sanitizeCollectionID");
const _findVersionByID = /*#__PURE__*/ _interop_require_default(require("../operations/findVersionByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function findVersionByIDHandler(req, res, next) {
    const id = (0, _sanitizeCollectionID.sanitizeCollectionID)({
        id: req.params.id,
        collectionSlug: req.collection.config.slug,
        payload: req.payload
    });
    const options = {
        id,
        collection: req.collection,
        depth: parseInt(String(req.query.depth), 10),
        payload: req.payload,
        req
    };
    try {
        const doc = await (0, _findVersionByID.default)(options);
        return res.json(doc);
    } catch (error) {
        return next(error);
    }
}
