// payload/dist/collections/requestHandlers/findByID.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findByIDHandler;
    }
});
const _sanitizeCollectionID = require("../../utilities/sanitizeCollectionID");
const _findByID = /*#__PURE__*/ _interop_require_default(require("../operations/findByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function findByIDHandler(req, res, next) {
    const id = (0, _sanitizeCollectionID.sanitizeCollectionID)({
        id: req.params.id,
        collectionSlug: req.collection.config.slug,
        payload: req.payload
    });
    try {
        const doc = await (0, _findByID.default)({
            id,
            collection: req.collection,
            depth: Number(req.query.depth),
            draft: req.query.draft === 'true',
            req
        });
        return res.json(doc);
    } catch (error) {
        return next(error);
    }
}
