// payload/dist/collections/requestHandlers/restoreVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return restoreVersionHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _formatSuccess = /*#__PURE__*/ _interop_require_default(require("../../express/responses/formatSuccess"));
const _sanitizeCollectionID = require("../../utilities/sanitizeCollectionID");
const _restoreVersion = /*#__PURE__*/ _interop_require_default(require("../operations/restoreVersion"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function restoreVersionHandler(req, res, next) {
    const id = (0, _sanitizeCollectionID.sanitizeCollectionID)({
        id: req.params.id,
        collectionSlug: req.collection.config.slug,
        payload: req.payload
    });
    const options = {
        id,
        collection: req.collection,
        depth: Number(req.query.depth),
        payload: req.payload,
        req
    };
    try {
        const doc = await (0, _restoreVersion.default)(options);
        res.status(_httpstatus.default.OK).json({
            ...(0, _formatSuccess.default)(req.t('version:restoredSuccessfully'), 'message'),
            doc
        });
    } catch (error) {
        next(error);
    }
}
