// payload/dist/collections/requestHandlers/update.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return updateHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _formatSuccess = /*#__PURE__*/ _interop_require_default(require("../../express/responses/formatSuccess"));
const _getTranslation = require("../../utilities/getTranslation");
const _update = /*#__PURE__*/ _interop_require_default(require("../operations/update"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function updateHandler(req, res, next) {
    try {
        const draft = req.query.draft === 'true';
        const result = await (0, _update.default)({
            collection: req.collection,
            data: req.body,
            depth: parseInt(String(req.query.depth), 10),
            draft,
            req,
            where: req.query.where
        });
        if (result.errors.length === 0) {
            const message = req.t('general:updatedCountSuccessfully', {
                count: result.docs.length,
                label: (0, _getTranslation.getTranslation)(req.collection.config.labels[result.docs.length > 1 ? 'plural' : 'singular'], req.i18n)
            });
            res.status(_httpstatus.default.OK).json({
                ...(0, _formatSuccess.default)(message, 'message'),
                ...result
            });
            return;
        }
        const total = result.docs.length + result.errors.length;
        const message = req.t('error:unableToUpdateCount', {
            count: result.errors.length,
            label: (0, _getTranslation.getTranslation)(req.collection.config.labels[total > 1 ? 'plural' : 'singular'], req.i18n),
            total
        });
        res.status(_httpstatus.default.BAD_REQUEST).json({
            ...(0, _formatSuccess.default)(message, 'message'),
            ...result
        });
    } catch (error) {
        next(error);
    }
}
