// payload/dist/admin/components/views/Version/RenderFieldsToDiff/fields/Tabs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _ = /*#__PURE__*/ _interop_require_default(require("../.."));
const _Nested = /*#__PURE__*/ _interop_require_default(require("../Nested"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'tabs-diff';
const Tabs = ({ comparison, field, fieldComponents, locales, permissions, version })=>/*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrap`
    }, field.tabs.map((tab, i)=>{
        if ('name' in tab) {
            return /*#__PURE__*/ _react.default.createElement(_Nested.default, {
                comparison: comparison?.[tab.name],
                field: tab,
                fieldComponents: fieldComponents,
                key: i,
                locales: locales,
                permissions: permissions,
                version: version?.[tab.name]
            });
        }
        return /*#__PURE__*/ _react.default.createElement(_.default, {
            comparison: comparison,
            fieldComponents: fieldComponents,
            fieldPermissions: permissions,
            fields: tab.fields,
            key: i,
            locales: locales,
            version: version
        });
    })));
const _default = Tabs;
