// payload/dist/database/createDatabaseAdapter.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "createDatabaseAdapter", {
    enumerable: true,
    get: function() {
        return createDatabaseAdapter;
    }
});
const _createMigration = require("./migrations/createMigration");
const _migrate = require("./migrations/migrate");
const _migrateDown = require("./migrations/migrateDown");
const _migrateRefresh = require("./migrations/migrateRefresh");
const _migrateReset = require("./migrations/migrateReset");
const _migrateStatus = require("./migrations/migrateStatus");
const beginTransaction = async ()=>null;
const rollbackTransaction = async ()=>null;
const commitTransaction = async ()=>null;
function createDatabaseAdapter(args) {
    return {
        // Default 'null' transaction functions
        beginTransaction,
        commitTransaction,
        createMigration: _createMigration.createMigration,
        migrate: _migrate.migrate,
        migrateDown: _migrateDown.migrateDown,
        migrateFresh: async ({ forceAcceptWarning = null })=>null,
        migrateRefresh: _migrateRefresh.migrateRefresh,
        migrateReset: _migrateReset.migrateReset,
        migrateStatus: _migrateStatus.migrateStatus,
        rollbackTransaction,
        ...args,
        // Ensure migrationDir is set
        migrationDir: args.migrationDir || 'migrations'
    };
}
