// payload/dist/auth/operations/logout.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _errors = require("../../errors");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function logout(incomingArgs) {
    let args = incomingArgs;
    const { collection, collection: { config: collectionConfig }, req, req: { payload: { config }, user }, res } = incomingArgs;
    if (!user) throw new _errors.APIError('No User', _httpstatus.default.BAD_REQUEST);
    if (user.collection !== collectionConfig.slug) throw new _errors.APIError('Incorrect collection', _httpstatus.default.FORBIDDEN);
    const cookieOptions = {
        domain: undefined,
        httpOnly: true,
        path: '/',
        sameSite: collectionConfig.auth.cookies.sameSite,
        secure: collectionConfig.auth.cookies.secure
    };
    if (collectionConfig.auth.cookies.domain) cookieOptions.domain = collectionConfig.auth.cookies.domain;
    await collection.config.hooks.afterLogout.reduce(async (priorHook, hook)=>{
        await priorHook;
        args = await hook({
            collection: args.collection?.config,
            context: req.context,
            req,
            res
        }) || args;
    }, Promise.resolve());
    res.clearCookie(`${config.cookiePrefix}-token`, cookieOptions);
    return req.t('authentication:loggedOutSuccessfully');
}
const _default = logout;
