// payload/dist/auth/operations/me.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _jsonwebtoken = /*#__PURE__*/ _interop_require_default(require("jsonwebtoken"));
const _url = /*#__PURE__*/ _interop_require_default(require("url"));
const _getExtractJWT = /*#__PURE__*/ _interop_require_default(require("../getExtractJWT"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function me(args) {
    const { collection, req } = args;
    const extractJWT = (0, _getExtractJWT.default)(req.payload.config);
    let response = {
        user: null
    };
    if (req.user) {
        const parsedURL = _url.default.parse(req.originalUrl);
        const isGraphQL = parsedURL.pathname === `/api${req.payload.config.routes.graphQL}`;
        const user = await req.payload.findByID({
            id: req.user.id,
            collection: collection.config.slug,
            depth: isGraphQL ? 0 : collection.config.auth.depth,
            overrideAccess: false,
            req,
            showHiddenFields: false
        });
        if (req.user.collection !== collection.config.slug) {
            return {
                user: null
            };
        }
        delete user.collection;
        // /////////////////////////////////////
        // me hook - Collection
        // /////////////////////////////////////
        for (const meHook of collection.config.hooks.me){
            const hookResult = await meHook({
                args,
                user
            });
            if (hookResult) {
                response.user = hookResult.user;
                response.exp = hookResult.exp;
                break;
            }
        }
        response.collection = req.user.collection;
        response.strategy = req.user._strategy;
        const token = extractJWT(req);
        if (!response.user) {
            response.user = user;
            if (token) {
                const decoded = _jsonwebtoken.default.decode(token);
                if (decoded) response.exp = decoded.exp;
                if (!collection.config.auth.removeTokenFromResponses) response.token = token;
            }
        }
    }
    // /////////////////////////////////////
    // After Me - Collection
    // /////////////////////////////////////
    await collection.config.hooks.afterMe.reduce(async (priorHook, hook)=>{
        await priorHook;
        response = await hook({
            collection: collection?.config,
            context: req.context,
            req,
            response
        }) || response;
    }, Promise.resolve());
    return response;
}
const _default = me;
