// payload/dist/auth/operations/registerFirstUser.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _errors = require("../../errors");
const _commitTransaction = require("../../utilities/commitTransaction");
const _initTransaction = require("../../utilities/initTransaction");
const _killTransaction = require("../../utilities/killTransaction");
async function registerFirstUser(args) {
    const { collection: { config, config: { auth: { verify }, slug } }, data, req, req: { payload } } = args;
    try {
        const shouldCommit = await (0, _initTransaction.initTransaction)(req);
        const doc = await payload.db.findOne({
            collection: config.slug,
            req
        });
        if (doc) throw new _errors.Forbidden(req.t);
        // /////////////////////////////////////
        // Register first user
        // /////////////////////////////////////
        const result = await payload.create({
            collection: slug,
            data,
            overrideAccess: true,
            req
        });
        // auto-verify (if applicable)
        if (verify) {
            await payload.update({
                id: result.id,
                collection: slug,
                data: {
                    _verified: true
                },
                req
            });
        }
        // /////////////////////////////////////
        // Log in new user
        // /////////////////////////////////////
        const { token } = await payload.login({
            ...args,
            collection: slug,
            req
        });
        const resultToReturn = {
            ...result,
            token
        };
        if (shouldCommit) await (0, _commitTransaction.commitTransaction)(req);
        return {
            message: 'Registered and logged in successfully. Welcome!',
            user: resultToReturn
        };
    } catch (error) {
        await (0, _killTransaction.killTransaction)(req);
        throw error;
    }
}
const _default = registerFirstUser;
