// payload/dist/auth/operations/unlock.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _errors = require("../../errors");
const _commitTransaction = require("../../utilities/commitTransaction");
const _initTransaction = require("../../utilities/initTransaction");
const _killTransaction = require("../../utilities/killTransaction");
const _executeAccess = /*#__PURE__*/ _interop_require_default(require("../executeAccess"));
const _resetLoginAttempts = require("../strategies/local/resetLoginAttempts");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function unlock(args) {
    if (!Object.prototype.hasOwnProperty.call(args.data, 'email')) {
        throw new _errors.APIError('Missing email.');
    }
    const { collection: { config: collectionConfig }, overrideAccess, req: { locale, payload }, req } = args;
    try {
        const shouldCommit = await (0, _initTransaction.initTransaction)(req);
        // /////////////////////////////////////
        // Access
        // /////////////////////////////////////
        if (!overrideAccess) {
            await (0, _executeAccess.default)({
                req
            }, collectionConfig.access.unlock);
        }
        const options = {
            ...args
        };
        const { data } = options;
        // /////////////////////////////////////
        // Unlock
        // /////////////////////////////////////
        if (!data.email) {
            throw new _errors.APIError('Missing email.');
        }
        const user = await req.payload.db.findOne({
            collection: collectionConfig.slug,
            locale,
            req,
            where: {
                email: {
                    equals: data.email.toLowerCase()
                }
            }
        });
        let result;
        if (user) {
            await (0, _resetLoginAttempts.resetLoginAttempts)({
                collection: collectionConfig,
                doc: user,
                payload: req.payload,
                req
            });
            result = true;
        } else {
            result = null;
        }
        if (shouldCommit) await (0, _commitTransaction.commitTransaction)(req);
        return result;
    } catch (error) {
        await (0, _killTransaction.killTransaction)(req);
        throw error;
    }
}
const _default = unlock;
