// payload/dist/admin/components/forms/field-types/ConfirmPassword/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _Error = /*#__PURE__*/ _interop_require_default(require("../../Error"));
const _context = require("../../Form/context");
const _Label = /*#__PURE__*/ _interop_require_default(require("../../Label"));
const _useField = /*#__PURE__*/ _interop_require_default(require("../../useField"));
require("./index.scss");
const _shared = require("../shared");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ConfirmPassword = (props)=>{
    const { disabled } = props;
    const password = (0, _context.useFormFields)(([fields])=>fields.password);
    const { t } = (0, _reacti18next.useTranslation)('fields');
    const validate = (0, _react.useCallback)((value)=>{
        if (!value) {
            return t('validation:required');
        }
        if (value === password?.value) {
            return true;
        }
        return t('passwordsDoNotMatch');
    }, [
        password,
        t
    ]);
    const { errorMessage, setValue, showError, value } = (0, _useField.default)({
        disableFormData: true,
        path: 'confirm-password',
        validate
    });
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            _shared.fieldBaseClass,
            'confirm-password',
            showError && 'error'
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_Error.default, {
        message: errorMessage,
        showError: showError
    }), /*#__PURE__*/ _react.default.createElement(_Label.default, {
        htmlFor: "field-confirm-password",
        label: t('authentication:confirmPassword'),
        required: true
    }), /*#__PURE__*/ _react.default.createElement("input", {
        autoComplete: "off",
        disabled: !!disabled,
        id: "field-confirm-password",
        name: "confirm-password",
        onChange: setValue,
        type: "password",
        value: value || ''
    }));
};
const _default = ConfirmPassword;
