// payload/dist/admin/components/views/collections/List/Cell/field-types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _Array = /*#__PURE__*/ _interop_require_default(require("./Array"));
const _Blocks = /*#__PURE__*/ _interop_require_default(require("./Blocks"));
const _Checkbox = /*#__PURE__*/ _interop_require_default(require("./Checkbox"));
const _Code = /*#__PURE__*/ _interop_require_default(require("./Code"));
const _Date = /*#__PURE__*/ _interop_require_default(require("./Date"));
const _File = /*#__PURE__*/ _interop_require_default(require("./File"));
const _JSON = /*#__PURE__*/ _interop_require_default(require("./JSON"));
const _Relationship = /*#__PURE__*/ _interop_require_default(require("./Relationship"));
const _Richtext = /*#__PURE__*/ _interop_require_default(require("./Richtext"));
const _Select = /*#__PURE__*/ _interop_require_default(require("./Select"));
const _Textarea = /*#__PURE__*/ _interop_require_default(require("./Textarea"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const _default = {
    File: _File.default,
    array: _Array.default,
    blocks: _Blocks.default,
    checkbox: _Checkbox.default,
    code: _Code.default,
    date: _Date.default,
    json: _JSON.default,
    radio: _Select.default,
    relationship: _Relationship.default,
    richText: _Richtext.default,
    select: _Select.default,
    textarea: _Textarea.default,
    upload: _Relationship.default
};
