// payload/dist/admin/components/views/LivePreview/Device/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DeviceContainer", {
    enumerable: true,
    get: function() {
        return DeviceContainer;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _useResize = require("../../../../utilities/useResize");
const _context = require("../Context/context");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const DeviceContainer = (props)=>{
    const { children } = props;
    const deviceFrameRef = _react.default.useRef(null);
    const outerFrameRef = _react.default.useRef(null);
    const { breakpoint, setMeasuredDeviceSize, size: desiredSize, zoom } = (0, _context.useLivePreviewContext)();
    // Keep an accurate measurement of the actual device size as it is truly rendered
    // This is helpful when `sizes` are non-number units like percentages, etc.
    const { size: measuredDeviceSize } = (0, _useResize.useResize)(deviceFrameRef.current);
    const { size: outerFrameSize } = (0, _useResize.useResize)(outerFrameRef.current);
    let deviceIsLargerThanFrame = false;
    // Sync the measured device size with the context so that other components can use it
    // This happens from the bottom up so that as this component mounts and unmounts,
    // its size is freshly populated again upon re-mounting, i.e. going from iframe->popup->iframe
    (0, _react.useEffect)(()=>{
        if (measuredDeviceSize) {
            setMeasuredDeviceSize(measuredDeviceSize);
        }
    }, [
        measuredDeviceSize,
        setMeasuredDeviceSize
    ]);
    let x = '0';
    let margin = '0';
    if (breakpoint && breakpoint !== 'responsive') {
        x = '-50%';
        if (typeof zoom === 'number' && typeof desiredSize.width === 'number' && typeof desiredSize.height === 'number' && typeof measuredDeviceSize.width === 'number' && typeof measuredDeviceSize.height === 'number') {
            margin = '0 auto';
            const scaledDesiredWidth = desiredSize.width / zoom;
            const scaledDeviceWidth = measuredDeviceSize.width * zoom;
            const scaledDeviceDifferencePixels = scaledDesiredWidth - desiredSize.width;
            deviceIsLargerThanFrame = scaledDeviceWidth > outerFrameSize.width;
            if (deviceIsLargerThanFrame) {
                if (zoom > 1) {
                    const differenceFromDeviceToFrame = measuredDeviceSize.width - outerFrameSize.width;
                    if (differenceFromDeviceToFrame < 0) x = `${differenceFromDeviceToFrame / 2}px`;
                    else x = '0';
                } else {
                    x = '0';
                }
            } else {
                if (zoom >= 1) {
                    x = `${scaledDeviceDifferencePixels / 2}px`;
                } else {
                    const differenceFromDeviceToFrame = outerFrameSize.width - scaledDeviceWidth;
                    x = `${differenceFromDeviceToFrame / 2}px`;
                    margin = '0';
                }
            }
        }
    }
    let width = zoom ? `${100 / zoom}%` : '100%';
    let height = zoom ? `${100 / zoom}%` : '100%';
    if (breakpoint !== 'responsive') {
        width = `${desiredSize?.width / (typeof zoom === 'number' ? zoom : 1)}px`;
        height = `${desiredSize?.height / (typeof zoom === 'number' ? zoom : 1)}px`;
    }
    return /*#__PURE__*/ _react.default.createElement("div", {
        ref: outerFrameRef,
        style: {
            height: '100%',
            width: '100%'
        }
    }, /*#__PURE__*/ _react.default.createElement("div", {
        ref: deviceFrameRef,
        style: {
            height,
            margin,
            transform: `translate3d(${x}, 0, 0)`,
            width
        }
    }, children));
};
