// payload/dist/email/mockHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _nodemailer = /*#__PURE__*/ _interop_require_default(require("nodemailer"));
const _defaults = require("./defaults");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const mockEmailHandler = async (emailConfig)=>{
    const testAccount = await _nodemailer.default.createTestAccount();
    const smtpOptions = {
        ...emailConfig,
        auth: {
            pass: testAccount.pass,
            user: testAccount.user
        },
        fromAddress: emailConfig?.fromAddress || _defaults.defaults.fromAddress,
        fromName: emailConfig?.fromName || _defaults.defaults.fromName,
        host: 'smtp.ethereal.email',
        port: 587,
        secure: false
    };
    return {
        account: testAccount,
        transport: _nodemailer.default.createTransport(smtpOptions)
    };
};
const _default = mockEmailHandler;
