// payload/dist/admin/components/elements/Tooltip/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _useIntersect = /*#__PURE__*/ _interop_require_default(require("../../../hooks/useIntersect"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Tooltip = (props)=>{
    const { alignCaret = 'center', boundingRef, children, className, delay = 350, show: showFromProps = true } = props;
    const [show, setShow] = _react.default.useState(showFromProps);
    const [position, setPosition] = _react.default.useState('top');
    const getTitleAttribute = (content)=>typeof content === 'string' ? content : '';
    const [ref, intersectionEntry] = (0, _useIntersect.default)({
        root: boundingRef?.current || null,
        rootMargin: '-145px 0px 0px 100px',
        threshold: 0
    });
    (0, _react.useEffect)(()=>{
        let timerId;
        // do not use the delay on transition-out
        if (delay && showFromProps) {
            timerId = setTimeout(()=>{
                setShow(showFromProps);
            }, delay);
        } else {
            setShow(showFromProps);
        }
        return ()=>{
            if (timerId) clearTimeout(timerId);
        };
    }, [
        showFromProps,
        delay
    ]);
    (0, _react.useEffect)(()=>{
        setPosition(intersectionEntry?.isIntersecting ? 'top' : 'bottom');
    }, [
        intersectionEntry
    ]);
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("aside", {
        "aria-hidden": "true",
        className: [
            'tooltip',
            className,
            `tooltip--caret-${alignCaret}`,
            'tooltip--position-top'
        ].filter(Boolean).join(' '),
        ref: ref,
        title: getTitleAttribute(children)
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: "tooltip-content"
    }, children)), /*#__PURE__*/ _react.default.createElement("aside", {
        className: [
            'tooltip',
            className,
            show && 'tooltip--show',
            `tooltip--caret-${alignCaret}`,
            `tooltip--position-${position}`
        ].filter(Boolean).join(' '),
        title: getTitleAttribute(children)
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: "tooltip-content"
    }, children)));
};
const _default = Tooltip;
