// payload/dist/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    APIError: function() {
        return _APIError.default;
    },
    AuthenticationError: function() {
        return _AuthenticationError.default;
    },
    DuplicateCollection: function() {
        return _DuplicateCollection.default;
    },
    DuplicateFieldName: function() {
        return _DuplicateFieldName.default;
    },
    DuplicateGlobal: function() {
        return _DuplicateGlobal.default;
    },
    ErrorDeletingFile: function() {
        return _ErrorDeletingFile.default;
    },
    FileUploadError: function() {
        return _FileUploadError.default;
    },
    Forbidden: function() {
        return _Forbidden.default;
    },
    InvalidConfiguration: function() {
        return _InvalidConfiguration.default;
    },
    InvalidFieldName: function() {
        return _InvalidFieldName.default;
    },
    InvalidFieldRelationship: function() {
        return _InvalidFieldRelationship.default;
    },
    LockedAuth: function() {
        return _LockedAuth.default;
    },
    MissingCollectionLabel: function() {
        return _MissingCollectionLabel.default;
    },
    MissingFieldInputOptions: function() {
        return _MissingFieldInputOptions.default;
    },
    MissingFieldType: function() {
        return _MissingFieldType.default;
    },
    MissingFile: function() {
        return _MissingFile.default;
    },
    NotFound: function() {
        return _NotFound.default;
    },
    QueryError: function() {
        return _QueryError.default;
    },
    ValidationError: function() {
        return _ValidationError.default;
    },
    errorHandler: function() {
        return _errorHandler.default;
    }
});
const _errorHandler = /*#__PURE__*/ _interop_require_default(require("../express/middleware/errorHandler"));
const _APIError = /*#__PURE__*/ _interop_require_default(require("./APIError"));
const _AuthenticationError = /*#__PURE__*/ _interop_require_default(require("./AuthenticationError"));
const _DuplicateCollection = /*#__PURE__*/ _interop_require_default(require("./DuplicateCollection"));
const _DuplicateFieldName = /*#__PURE__*/ _interop_require_default(require("./DuplicateFieldName"));
const _DuplicateGlobal = /*#__PURE__*/ _interop_require_default(require("./DuplicateGlobal"));
const _ErrorDeletingFile = /*#__PURE__*/ _interop_require_default(require("./ErrorDeletingFile"));
const _FileUploadError = /*#__PURE__*/ _interop_require_default(require("./FileUploadError"));
const _Forbidden = /*#__PURE__*/ _interop_require_default(require("./Forbidden"));
const _InvalidConfiguration = /*#__PURE__*/ _interop_require_default(require("./InvalidConfiguration"));
const _InvalidFieldName = /*#__PURE__*/ _interop_require_default(require("./InvalidFieldName"));
const _InvalidFieldRelationship = /*#__PURE__*/ _interop_require_default(require("./InvalidFieldRelationship"));
const _LockedAuth = /*#__PURE__*/ _interop_require_default(require("./LockedAuth"));
const _MissingCollectionLabel = /*#__PURE__*/ _interop_require_default(require("./MissingCollectionLabel"));
const _MissingFieldInputOptions = /*#__PURE__*/ _interop_require_default(require("./MissingFieldInputOptions"));
const _MissingFieldType = /*#__PURE__*/ _interop_require_default(require("./MissingFieldType"));
const _MissingFile = /*#__PURE__*/ _interop_require_default(require("./MissingFile"));
const _NotFound = /*#__PURE__*/ _interop_require_default(require("./NotFound"));
const _QueryError = /*#__PURE__*/ _interop_require_default(require("./QueryError"));
const _ValidationError = /*#__PURE__*/ _interop_require_default(require("./ValidationError"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
