// payload/dist/admin/components/views/collections/List/formatFields.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _types = require("../../../../../fields/config/types");
const formatFields = (config)=>{
    const hasID = config.fields.findIndex((field)=>(0, _types.fieldAffectsData)(field) && field.name === 'id') > -1;
    const defaultIDField = {
        name: 'id',
        type: 'text',
        admin: {
            disableBulkEdit: true
        },
        label: 'ID'
    };
    const shouldSkipField = (field)=>!(0, _types.fieldIsPresentationalOnly)(field) && (field.hidden === true || field.admin?.disabled === true);
    const fields = config.fields.reduce((formatted, field)=>{
        if (shouldSkipField(field)) {
            return formatted;
        }
        const formattedField = field.type === 'tabs' ? {
            ...field,
            tabs: field.tabs.map((tab)=>({
                    ...tab,
                    fields: tab.fields.filter((tabField)=>!shouldSkipField(tabField))
                }))
        } : field;
        return [
            ...formatted,
            formattedField
        ];
    }, hasID ? [] : [
        defaultIDField
    ]);
    return fields;
};
const _default = formatFields;
