// payload/dist/admin/components/forms/field-types/Upload/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _validations = require("../../../../../fields/validations");
const _Config = require("../../../utilities/Config");
const _useField = /*#__PURE__*/ _interop_require_default(require("../../useField"));
const _withCondition = /*#__PURE__*/ _interop_require_default(require("../../withCondition"));
const _Input = /*#__PURE__*/ _interop_require_default(require("./Input"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Upload = (props)=>{
    const { collections, routes: { api }, serverURL } = (0, _Config.useConfig)();
    const { name, admin: { className, condition, description, readOnly, style, width, components: { Error, Label } = {} } = {}, fieldTypes, filterOptions, label, path, relationTo, required, validate = _validations.upload } = props;
    const collection = collections.find((coll)=>coll.slug === relationTo);
    const memoizedValidate = (0, _react.useCallback)((value, options)=>{
        return validate(value, {
            ...options,
            required
        });
    }, [
        validate,
        required
    ]);
    const field = (0, _useField.default)({
        condition,
        path,
        validate: memoizedValidate
    });
    const { errorMessage, setValue, showError, value } = field;
    const onChange = (0, _react.useCallback)((incomingValue)=>{
        const incomingID = incomingValue?.id || incomingValue;
        setValue(incomingID);
    }, [
        setValue
    ]);
    if (collection.upload) {
        return /*#__PURE__*/ _react.default.createElement(_Input.default, {
            api: api,
            className: className,
            collection: collection,
            description: description,
            errorMessage: errorMessage,
            fieldTypes: fieldTypes,
            filterOptions: filterOptions,
            label: label,
            name: name,
            onChange: onChange,
            path: path,
            readOnly: readOnly,
            relationTo: relationTo,
            required: required,
            serverURL: serverURL,
            showError: showError,
            style: style,
            value: value,
            width: width,
            Error: Error,
            Label: Label
        });
    }
    return null;
};
const _default = (0, _withCondition.default)(Upload);
