// payload/dist/admin/components/elements/FileDetails/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _isImage = /*#__PURE__*/ _interop_require_default(require("../../../../uploads/isImage"));
const _Upload = require("../../views/collections/Edit/Upload");
const _Button = /*#__PURE__*/ _interop_require_default(require("../Button"));
const _Thumbnail = /*#__PURE__*/ _interop_require_default(require("../Thumbnail"));
const _Meta = /*#__PURE__*/ _interop_require_default(require("./Meta"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'file-details';
const FileDetails = (props)=>{
    const { canEdit, collection, doc, handleRemove, hasImageSizes, imageCacheTag } = props;
    const { slug: collectionSlug, upload: { staticURL } } = collection;
    const { id, filename, filesize, height, mimeType, url, width } = doc;
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("header", null, /*#__PURE__*/ _react.default.createElement(_Thumbnail.default, {
        collection: collection,
        doc: doc,
        imageCacheTag: imageCacheTag
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__main-detail`
    }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        collection: collectionSlug,
        filename: filename,
        filesize: filesize,
        height: height,
        id: id,
        mimeType: mimeType,
        staticURL: staticURL,
        url: url,
        width: width
    }), (0, _isImage.default)(mimeType) && mimeType !== 'image/svg+xml' && /*#__PURE__*/ _react.default.createElement(_Upload.UploadActions, {
        canEdit: canEdit,
        showSizePreviews: hasImageSizes && doc.filename
    })), handleRemove && /*#__PURE__*/ _react.default.createElement(_Button.default, {
        buttonStyle: "icon-label",
        className: `${baseClass}__remove`,
        icon: "x",
        iconStyle: "with-border",
        onClick: handleRemove,
        round: true
    })));
};
const _default = FileDetails;
