// payload/dist/admin/components/elements/DraggableSortable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _core = require("@dnd-kit/core");
const _sortable = require("@dnd-kit/sortable");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const DraggableSortable = (props)=>{
    const { children, className, ids, onDragEnd } = props;
    const id = (0, _react.useId)();
    const { setNodeRef } = (0, _core.useDroppable)({
        id
    });
    const sensors = (0, _core.useSensors)((0, _core.useSensor)(_core.PointerSensor, {
        activationConstraint: {
            distance: 5
        }
    }), (0, _core.useSensor)(_core.KeyboardSensor, {
        coordinateGetter: _sortable.sortableKeyboardCoordinates
    }));
    const handleDragEnd = (0, _react.useCallback)((event)=>{
        const { active, over } = event;
        if (!active || !over) return;
        if (typeof onDragEnd === 'function') {
            onDragEnd({
                event,
                moveFromIndex: ids.findIndex((_id)=>_id === active.id),
                moveToIndex: ids.findIndex((_id)=>_id === over.id)
            });
        }
    }, [
        onDragEnd,
        ids
    ]);
    return /*#__PURE__*/ _react.default.createElement(_core.DndContext, {
        collisionDetection: _core.closestCenter,
        onDragEnd: handleDragEnd,
        sensors: sensors
    }, /*#__PURE__*/ _react.default.createElement(_sortable.SortableContext, {
        items: ids
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: className,
        ref: setNodeRef
    }, children)));
};
const _default = DraggableSortable;
