// payload/dist/admin/components/views/Account/Settings/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Settings", {
    enumerable: true,
    get: function() {
        return Settings;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _ReactSelect = /*#__PURE__*/ _interop_require_default(require("../../../elements/ReactSelect"));
const _Label = /*#__PURE__*/ _interop_require_default(require("../../../forms/Label"));
const _ToggleTheme = require("../ToggleTheme");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'payload-settings';
const Settings = (props)=>{
    const { className } = props;
    const { i18n, t } = (0, _reacti18next.useTranslation)('authentication');
    const languageOptions = Object.entries(i18n.options.resources || {}).map(([language, resource])=>({
            label: resource.general.thisLanguage,
            value: language
        }));
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            className
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement("h3", null, t('general:payloadSettings')), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__language`
    }, /*#__PURE__*/ _react.default.createElement(_Label.default, {
        htmlFor: "language-select",
        label: t('general:language')
    }), /*#__PURE__*/ _react.default.createElement(_ReactSelect.default, {
        inputId: "language-select",
        onChange: ({ value })=>i18n.changeLanguage(value),
        options: languageOptions,
        value: languageOptions.find((language)=>language.value === i18n.language)
    })), /*#__PURE__*/ _react.default.createElement(_ToggleTheme.ToggleTheme, null));
};
