// payload/dist/admin/components/elements/PreviewButton/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reacttoastify = require("react-toastify");
const _Auth = require("../../utilities/Auth");
const _Config = require("../../utilities/Config");
const _DocumentInfo = require("../../utilities/DocumentInfo");
const _Locale = require("../../utilities/Locale");
const _RenderCustomComponent = /*#__PURE__*/ _interop_require_default(require("../../utilities/RenderCustomComponent"));
const _Button = /*#__PURE__*/ _interop_require_default(require("../Button"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'preview-btn';
const DefaultPreviewButton = ({ disabled, label, preview })=>{
    return /*#__PURE__*/ _react.default.createElement(_Button.default, {
        buttonStyle: "secondary",
        className: baseClass,
        disabled: disabled,
        onClick: preview,
        size: "small"
    }, label);
};
const PreviewButton = ({ CustomComponent, generatePreviewURL })=>{
    const { id, collection, global } = (0, _DocumentInfo.useDocumentInfo)();
    const [isLoading, setIsLoading] = (0, _react.useState)(false);
    const { code: locale } = (0, _Locale.useLocale)();
    const { token } = (0, _Auth.useAuth)();
    const { routes: { api }, serverURL } = (0, _Config.useConfig)();
    const { t } = (0, _reacti18next.useTranslation)('version');
    const isGeneratingPreviewURL = (0, _react.useRef)(false);
    // we need to regenerate the preview URL every time the button is clicked
    // to do this we need to fetch the document data fresh from the API
    // this will ensure the latest data is used when generating the preview URL
    const preview = (0, _react.useCallback)(async ()=>{
        if (!generatePreviewURL || isGeneratingPreviewURL.current) return;
        isGeneratingPreviewURL.current = true;
        try {
            setIsLoading(true);
            let url = `${serverURL}${api}`;
            if (collection) url = `${url}/${collection.slug}/${id}`;
            if (global) url = `${url}/globals/${global.slug}`;
            const data = await fetch(`${url}?draft=true&locale=${locale}&fallback-locale=null`).then((res)=>res.json());
            const previewURL = await generatePreviewURL(data, {
                locale,
                token
            });
            if (!previewURL) throw new Error();
            setIsLoading(false);
            isGeneratingPreviewURL.current = false;
            window.open(previewURL, '_blank');
        } catch (err) {
            setIsLoading(false);
            isGeneratingPreviewURL.current = false;
            _reacttoastify.toast.error(t('error:previewing'));
        }
    }, [
        serverURL,
        api,
        collection,
        global,
        id,
        generatePreviewURL,
        locale,
        token,
        t
    ]);
    return /*#__PURE__*/ _react.default.createElement(_RenderCustomComponent.default, {
        CustomComponent: CustomComponent,
        DefaultComponent: DefaultPreviewButton,
        componentProps: {
            DefaultButton: DefaultPreviewButton,
            disabled: isLoading || !generatePreviewURL,
            label: isLoading ? t('general:loading') : t('preview'),
            preview
        }
    });
};
const _default = PreviewButton;
