// payload/dist/admin/components/forms/field-types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "fieldTypes", {
    enumerable: true,
    get: function() {
        return fieldTypes;
    }
});
const _Array = /*#__PURE__*/ _interop_require_default(require("./Array"));
const _Blocks = /*#__PURE__*/ _interop_require_default(require("./Blocks"));
const _Checkbox = /*#__PURE__*/ _interop_require_default(require("./Checkbox"));
const _Code = /*#__PURE__*/ _interop_require_default(require("./Code"));
const _Collapsible = /*#__PURE__*/ _interop_require_default(require("./Collapsible"));
const _ConfirmPassword = /*#__PURE__*/ _interop_require_default(require("./ConfirmPassword"));
const _DateTime = /*#__PURE__*/ _interop_require_default(require("./DateTime"));
const _Email = /*#__PURE__*/ _interop_require_default(require("./Email"));
const _Group = /*#__PURE__*/ _interop_require_default(require("./Group"));
const _HiddenInput = /*#__PURE__*/ _interop_require_default(require("./HiddenInput"));
const _JSON = /*#__PURE__*/ _interop_require_default(require("./JSON"));
const _Number = /*#__PURE__*/ _interop_require_default(require("./Number"));
const _Password = /*#__PURE__*/ _interop_require_default(require("./Password"));
const _Point = /*#__PURE__*/ _interop_require_default(require("./Point"));
const _RadioGroup = /*#__PURE__*/ _interop_require_default(require("./RadioGroup"));
const _Relationship = /*#__PURE__*/ _interop_require_default(require("./Relationship"));
const _RichText = /*#__PURE__*/ _interop_require_default(require("./RichText"));
const _Row = /*#__PURE__*/ _interop_require_default(require("./Row"));
const _Select = /*#__PURE__*/ _interop_require_default(require("./Select"));
const _Tabs = /*#__PURE__*/ _interop_require_default(require("./Tabs"));
const _Text = /*#__PURE__*/ _interop_require_default(require("./Text"));
const _Textarea = /*#__PURE__*/ _interop_require_default(require("./Textarea"));
const _UI = /*#__PURE__*/ _interop_require_default(require("./UI"));
const _Upload = /*#__PURE__*/ _interop_require_default(require("./Upload"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const fieldTypes = {
    array: _Array.default,
    blocks: _Blocks.default,
    checkbox: _Checkbox.default,
    code: _Code.default,
    collapsible: _Collapsible.default,
    confirmPassword: _ConfirmPassword.default,
    date: _DateTime.default,
    email: _Email.default,
    group: _Group.default,
    hidden: _HiddenInput.default,
    json: _JSON.default,
    number: _Number.default,
    password: _Password.default,
    point: _Point.default,
    radio: _RadioGroup.default,
    relationship: _Relationship.default,
    richText: _RichText.default,
    row: _Row.default,
    select: _Select.default,
    tabs: _Tabs.default,
    text: _Text.default,
    textarea: _Textarea.default,
    ui: _UI.default,
    upload: _Upload.default
};
