// payload/dist/admin/components/forms/field-types/Relationship/select-components/SingleValue/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "SingleValue", {
    enumerable: true,
    get: function() {
        return SingleValue;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactselect = require("react-select");
const _DocumentDrawer = require("../../../../../elements/DocumentDrawer");
const _Tooltip = /*#__PURE__*/ _interop_require_default(require("../../../../../elements/Tooltip"));
const _Edit = /*#__PURE__*/ _interop_require_default(require("../../../../../icons/Edit"));
const _Auth = require("../../../../../utilities/Auth");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'relationship--single-value';
const SingleValue = (props)=>{
    const { children, data: { label, relationTo, value }, // @ts-expect-error // TODO Fix this - moduleResolution 16 breaks our declare module
    selectProps: { customProps: { onSave, setDrawerIsOpen } = {} } = {} } = props;
    const [showTooltip, setShowTooltip] = (0, _react.useState)(false);
    const { t } = (0, _reacti18next.useTranslation)('general');
    const { permissions } = (0, _Auth.useAuth)();
    const hasReadPermission = Boolean(permissions?.collections?.[relationTo]?.read?.permission);
    const [DocumentDrawer, DocumentDrawerToggler, { isDrawerOpen }] = (0, _DocumentDrawer.useDocumentDrawer)({
        id: value.toString(),
        collectionSlug: relationTo
    });
    (0, _react.useEffect)(()=>{
        if (typeof setDrawerIsOpen === 'function') {
            setDrawerIsOpen(isDrawerOpen);
        }
    }, [
        isDrawerOpen,
        setDrawerIsOpen
    ]);
    return /*#__PURE__*/ _react.default.createElement(_reactselect.components.SingleValue, {
        ...props,
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__label`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__label-text`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__text`
    }, children), relationTo && hasReadPermission && /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, /*#__PURE__*/ _react.default.createElement(DocumentDrawerToggler, {
        "aria-label": t('editLabel', {
            label
        }),
        className: `${baseClass}__drawer-toggler`,
        onClick: ()=>setShowTooltip(false),
        onMouseDown: (e)=>e.stopPropagation(),
        onMouseEnter: ()=>setShowTooltip(true),
        onMouseLeave: ()=>setShowTooltip(false),
        onTouchEnd: (e)=>e.stopPropagation()
    }, /*#__PURE__*/ _react.default.createElement(_Tooltip.default, {
        className: `${baseClass}__tooltip`,
        show: showTooltip
    }, t('edit')), /*#__PURE__*/ _react.default.createElement(_Edit.default, null))))), relationTo && hasReadPermission && /*#__PURE__*/ _react.default.createElement(DocumentDrawer, {
        onSave: onSave
    }));
};
