// payload/dist/admin/components/views/Verify/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _Button = /*#__PURE__*/ _interop_require_default(require("../../elements/Button"));
const _Logo = /*#__PURE__*/ _interop_require_default(require("../../graphics/Logo"));
const _Minimal = /*#__PURE__*/ _interop_require_default(require("../../templates/Minimal"));
const _Auth = require("../../utilities/Auth");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
const _Login = /*#__PURE__*/ _interop_require_default(require("../Login"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'verify';
const Verify = ({ collection })=>{
    const { slug: collectionSlug } = collection;
    const { user } = (0, _Auth.useAuth)();
    const { token } = (0, _reactrouterdom.useParams)();
    const { admin: { user: adminUser }, routes: { admin: adminRoute }, serverURL } = (0, _Config.useConfig)();
    const { i18n, t } = (0, _reacti18next.useTranslation)('authentication');
    const isAdminUser = collectionSlug === adminUser;
    const [verifyResult, setVerifyResult] = (0, _react.useState)(null);
    (0, _react.useEffect)(()=>{
        async function verifyToken() {
            const result = await fetch(`${serverURL}/api/${collectionSlug}/verify/${token}`, {
                credentials: 'include',
                headers: {
                    'Accept-Language': i18n.language
                },
                method: 'POST'
            });
            setVerifyResult(result);
        }
        verifyToken();
    }, [
        setVerifyResult,
        collectionSlug,
        serverURL,
        token,
        i18n
    ]);
    if (user) {
        return /*#__PURE__*/ _react.default.createElement(_Login.default, null);
    }
    const getText = ()=>{
        if (verifyResult?.status === 200) return t('verifiedSuccessfully');
        if (verifyResult?.status === 202) return t('alreadyActivated');
        return t('unableToVerify');
    };
    return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: t('verifyUser'),
        keywords: t('verify'),
        title: t('verify')
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__brand`
    }, /*#__PURE__*/ _react.default.createElement(_Logo.default, null)), /*#__PURE__*/ _react.default.createElement("h2", null, getText()), isAdminUser && verifyResult?.status === 200 && /*#__PURE__*/ _react.default.createElement(_Button.default, {
        buttonStyle: "secondary",
        el: "link",
        to: `${adminRoute}/login`
    }, t('login')));
};
const _default = Verify;
