// payload/dist/auth/baseFields/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _validations = require("../../fields/validations");
const _extractTranslations = require("../../translations/extractTranslations");
const labels = (0, _extractTranslations.extractTranslations)([
    'general:email'
]);
const baseAuthFields = [
    {
        name: 'email',
        admin: {
            components: {
                Field: ()=>null
            }
        },
        label: labels['general:email'],
        required: true,
        type: 'email',
        unique: true,
        validate: _validations.email
    },
    {
        name: 'resetPasswordToken',
        hidden: true,
        type: 'text'
    },
    {
        name: 'resetPasswordExpiration',
        hidden: true,
        type: 'date'
    },
    {
        name: 'salt',
        hidden: true,
        type: 'text'
    },
    {
        name: 'hash',
        hidden: true,
        type: 'text'
    }
];
const _default = baseAuthFields;
