// payload/dist/auth/baseFields/verification.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _extractTranslations = require("../../translations/extractTranslations");
const labels = (0, _extractTranslations.extractTranslations)([
    'authentication:verified'
]);
const autoRemoveVerificationToken = ({ data, operation, originalDoc, value })=>{
    // If a user manually sets `_verified` to true,
    // and it was `false`, set _verificationToken to `null`.
    // This is useful because the admin panel
    // allows users to set `_verified` to true manually
    if (operation === 'update') {
        if (data?._verified === true && originalDoc?._verified === false) {
            return null;
        }
    }
    return value;
};
const _default = [
    {
        name: '_verified',
        access: {
            create: ({ req: { user } })=>Boolean(user),
            read: ({ req: { user } })=>Boolean(user),
            update: ({ req: { user } })=>Boolean(user)
        },
        admin: {
            components: {
                Field: ()=>null
            }
        },
        label: labels['authentication:verified'],
        type: 'checkbox'
    },
    {
        name: '_verificationToken',
        hidden: true,
        hooks: {
            beforeChange: [
                autoRemoveVerificationToken
            ]
        },
        type: 'text'
    }
];
