// payload/dist/utilities/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    default: function() {
        return _default;
    },
    defaultLoggerOptions: function() {
        return defaultLoggerOptions;
    },
    prettySyncLoggerDestination: function() {
        return prettySyncLoggerDestination;
    }
});
const _pino = /*#__PURE__*/ _interop_require_default(require("pino"));
const _pinopretty = /*#__PURE__*/ _interop_require_default(require("pino-pretty"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const prettyOptions = {
    colorize: true,
    ignore: 'pid,hostname',
    translateTime: 'SYS:HH:MM:ss'
};
const defaultLoggerOptions = {
    transport: {
        options: prettyOptions,
        target: 'pino-pretty'
    }
};
const prettySyncLoggerDestination = (0, _pinopretty.default)({
    ...prettyOptions,
    destination: 1,
    sync: true
});
const getLogger = (name = 'payload', options, destination)=>(0, _pino.default)({
        name: options?.name || name,
        enabled: process.env.DISABLE_LOGGING !== 'true',
        ...options || defaultLoggerOptions
    }, destination);
const _default = getLogger;
