// payload/dist/admin/components/elements/Popup/PopupTrigger/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PopupTrigger", {
    enumerable: true,
    get: function() {
        return PopupTrigger;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'popup-button';
const PopupTrigger = (props)=>{
    const { active, button, buttonType, className, setActive } = props;
    const classes = [
        baseClass,
        className,
        `${baseClass}--${buttonType}`
    ].filter(Boolean).join(' ');
    const handleClick = ()=>{
        setActive(!active);
    };
    if (buttonType === 'none') {
        return null;
    }
    if (buttonType === 'custom') {
        return /*#__PURE__*/ _react.default.createElement("div", {
            className: classes,
            onClick: handleClick,
            onKeyDown: (e)=>{
                if (e.key === 'Enter') handleClick();
            },
            role: "button",
            tabIndex: 0
        }, button);
    }
    return /*#__PURE__*/ _react.default.createElement("button", {
        className: classes,
        onClick: ()=>setActive(!active),
        tabIndex: 0,
        type: "button"
    }, button);
};
