// payload/dist/admin/components/forms/Form/context.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    FormContext: function() {
        return FormContext;
    },
    FormFieldsContext: function() {
        return FormFieldsContext;
    },
    FormWatchContext: function() {
        return FormWatchContext;
    },
    ModifiedContext: function() {
        return ModifiedContext;
    },
    ProcessingContext: function() {
        return ProcessingContext;
    },
    SubmittedContext: function() {
        return SubmittedContext;
    },
    useAllFormFields: function() {
        return useAllFormFields;
    },
    useForm: function() {
        return useForm;
    },
    useFormFields: function() {
        return useFormFields;
    },
    useFormModified: function() {
        return useFormModified;
    },
    useFormProcessing: function() {
        return useFormProcessing;
    },
    useFormSubmitted: function() {
        return useFormSubmitted;
    },
    useWatchForm: function() {
        return useWatchForm;
    }
});
const _react = require("react");
const _usecontextselector = require("use-context-selector");
const FormContext = (0, _react.createContext)({});
const FormWatchContext = (0, _react.createContext)({});
const SubmittedContext = (0, _react.createContext)(false);
const ProcessingContext = (0, _react.createContext)(false);
const ModifiedContext = (0, _react.createContext)(false);
const FormFieldsContext = (0, _usecontextselector.createContext)([
    {},
    ()=>null
]);
/**
 * Get the state of the form, can be used to submit & validate the form.
 *
 * @see https://payloadcms.com/docs/admin/hooks#useform
 */ const useForm = ()=>(0, _react.useContext)(FormContext);
const useWatchForm = ()=>(0, _react.useContext)(FormWatchContext);
const useFormSubmitted = ()=>(0, _react.useContext)(SubmittedContext);
const useFormProcessing = ()=>(0, _react.useContext)(ProcessingContext);
const useFormModified = ()=>(0, _react.useContext)(ModifiedContext);
/**
 * Get and set the value of a form field based on a selector
 *
 * @see https://payloadcms.com/docs/admin/hooks#useformfields
 */ const useFormFields = (selector)=>(0, _usecontextselector.useContextSelector)(FormFieldsContext, selector);
/**
 * Get the state of all form fields.
 *
 * @see https://payloadcms.com/docs/admin/hooks#useallformfields
 */ const useAllFormFields = ()=>(0, _usecontextselector.useContext)(FormFieldsContext);
