// payload/dist/admin/components/elements/ShimmerEffect/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ShimmerEffect: function() {
        return ShimmerEffect;
    },
    StaggeredShimmers: function() {
        return StaggeredShimmers;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _useDelay = require("../../../hooks/useDelay");
require("./index.scss");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ShimmerEffect = ({ animationDelay = '0ms', height = '60px', width = '100%' })=>{
    return /*#__PURE__*/ _react.createElement("div", {
        className: "shimmer-effect",
        style: {
            height: typeof height === 'number' ? `${height}px` : height,
            width: typeof width === 'number' ? `${width}px` : width
        }
    }, /*#__PURE__*/ _react.createElement("div", {
        className: "shimmer-effect__shine",
        style: {
            animationDelay
        }
    }));
};
const StaggeredShimmers = ({ className, count, height, renderDelay = 500, shimmerDelay = 25, shimmerItemClassName, width })=>{
    const shimmerDelayToPass = typeof shimmerDelay === 'number' ? `${shimmerDelay}ms` : shimmerDelay;
    const [hasDelayed] = (0, _useDelay.useDelay)(renderDelay, true);
    if (!hasDelayed) return null;
    return /*#__PURE__*/ _react.createElement("div", {
        className: className
    }, [
        ...Array(count)
    ].map((_, i)=>/*#__PURE__*/ _react.createElement("div", {
            className: shimmerItemClassName,
            key: i
        }, /*#__PURE__*/ _react.createElement(ShimmerEffect, {
            animationDelay: `calc(${i} * ${shimmerDelayToPass})`,
            height: height,
            width: width
        }))));
};
