// payload/dist/admin/components/elements/ListDrawer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ListDrawer: function() {
        return ListDrawer;
    },
    ListDrawerToggler: function() {
        return ListDrawerToggler;
    },
    baseClass: function() {
        return baseClass;
    },
    formatListDrawerSlug: function() {
        return formatListDrawerSlug;
    },
    useListDrawer: function() {
        return useListDrawer;
    }
});
const _modal = require("@faceless-ui/modal");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _Config = require("../../utilities/Config");
const _EditDepth = require("../../utilities/EditDepth");
const _Drawer = require("../Drawer");
const _DrawerContent = require("./DrawerContent");
require("./index.scss");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'list-drawer';
const formatListDrawerSlug = ({ depth, uuid })=>`list-drawer_${depth}_${uuid}`;
const ListDrawerToggler = ({ children, className, disabled, drawerSlug, ...rest })=>{
    return /*#__PURE__*/ _react.default.createElement(_Drawer.DrawerToggler, {
        className: [
            className,
            `${baseClass}__toggler`
        ].filter(Boolean).join(' '),
        disabled: disabled,
        slug: drawerSlug,
        ...rest
    }, children);
};
const ListDrawer = (props)=>{
    const { drawerSlug } = props;
    return /*#__PURE__*/ _react.default.createElement(_Drawer.Drawer, {
        className: baseClass,
        gutter: false,
        header: false,
        slug: drawerSlug
    }, /*#__PURE__*/ _react.default.createElement(_DrawerContent.ListDrawerContent, props));
};
const useListDrawer = ({ collectionSlugs: collectionSlugsFromProps, filterOptions, selectedCollection, uploads })=>{
    const { collections } = (0, _Config.useConfig)();
    const drawerDepth = (0, _EditDepth.useEditDepth)();
    const uuid = (0, _react.useId)();
    const { closeModal, modalState, openModal, toggleModal } = (0, _modal.useModal)();
    const [isOpen, setIsOpen] = (0, _react.useState)(false);
    const [collectionSlugs, setCollectionSlugs] = (0, _react.useState)(collectionSlugsFromProps);
    const drawerSlug = formatListDrawerSlug({
        depth: drawerDepth,
        uuid
    });
    (0, _react.useEffect)(()=>{
        setIsOpen(Boolean(modalState[drawerSlug]?.isOpen));
    }, [
        modalState,
        drawerSlug
    ]);
    (0, _react.useEffect)(()=>{
        if (!collectionSlugs || collectionSlugs.length === 0) {
            const filteredCollectionSlugs = collections.filter(({ upload })=>{
                if (uploads) {
                    return Boolean(upload) === true;
                }
                return true;
            });
            setCollectionSlugs(filteredCollectionSlugs.map(({ slug })=>slug));
        }
    }, [
        collectionSlugs,
        uploads,
        collections
    ]);
    const toggleDrawer = (0, _react.useCallback)(()=>{
        toggleModal(drawerSlug);
    }, [
        toggleModal,
        drawerSlug
    ]);
    const closeDrawer = (0, _react.useCallback)(()=>{
        closeModal(drawerSlug);
    }, [
        drawerSlug,
        closeModal
    ]);
    const openDrawer = (0, _react.useCallback)(()=>{
        openModal(drawerSlug);
    }, [
        drawerSlug,
        openModal
    ]);
    const MemoizedDrawer = (0, _react.useMemo)(()=>{
        return (props)=>/*#__PURE__*/ _react.default.createElement(ListDrawer, {
                ...props,
                closeDrawer: closeDrawer,
                collectionSlugs: collectionSlugs,
                drawerSlug: drawerSlug,
                filterOptions: filterOptions,
                key: drawerSlug,
                selectedCollection: selectedCollection,
                uploads: uploads
            });
    }, [
        drawerSlug,
        collectionSlugs,
        uploads,
        closeDrawer,
        selectedCollection,
        filterOptions
    ]);
    const MemoizedDrawerToggler = (0, _react.useMemo)(()=>{
        return (props)=>/*#__PURE__*/ _react.default.createElement(ListDrawerToggler, {
                ...props,
                drawerSlug: drawerSlug
            });
    }, [
        drawerSlug
    ]);
    const MemoizedDrawerState = (0, _react.useMemo)(()=>({
            closeDrawer,
            drawerDepth,
            drawerSlug,
            isDrawerOpen: isOpen,
            openDrawer,
            toggleDrawer
        }), [
        drawerDepth,
        drawerSlug,
        isOpen,
        toggleDrawer,
        closeDrawer,
        openDrawer
    ]);
    return [
        MemoizedDrawer,
        MemoizedDrawerToggler,
        MemoizedDrawerState
    ];
};
