// payload/dist/admin/components/forms/field-types/Blocks/SectionTitle/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _useField = /*#__PURE__*/ _interop_require_default(require("../../../useField"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'section-title';
/**
 * An input field representing the block's `blockName` property - responsible for reading and saving the `blockName`
 * property from/into the provided path.
 */ const SectionTitle = (props)=>{
    const { customOnChange, customValue, path, readOnly } = props;
    const { setValue, value } = (0, _useField.default)({
        path
    });
    const { t } = (0, _reacti18next.useTranslation)('general');
    const classes = [
        baseClass
    ].filter(Boolean).join(' ');
    const onChange = customOnChange || ((e)=>{
        e.stopPropagation();
        e.preventDefault();
        setValue(e.target.value);
    });
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: classes,
        "data-value": customValue || value
    }, /*#__PURE__*/ _react.default.createElement("input", {
        className: `${baseClass}__input`,
        id: path,
        name: path,
        onChange: onChange,
        placeholder: t('untitled'),
        readOnly: readOnly,
        type: "text",
        value: customValue || value || ''
    }));
};
const _default = SectionTitle;
