// payload/dist/admin/components/elements/Popup/PopupButtonList/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Button: function() {
        return Button;
    },
    ButtonGroup: function() {
        return ButtonGroup;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reactrouterdom = require("react-router-dom");
require("./index.scss");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'popup-button-list';
const ButtonGroup = ({ buttonSize = 'default', children, className, textAlign = 'left' })=>{
    const classes = [
        baseClass,
        className,
        `${baseClass}__text-align--${textAlign}`,
        `${baseClass}__button-size--${buttonSize}`
    ].filter(Boolean).join(' ');
    return /*#__PURE__*/ _react.createElement("div", {
        className: classes
    }, children);
};
const Button = ({ id, active, children, className, disabled, onClick, to })=>{
    const classes = [
        `${baseClass}__button`,
        active && `${baseClass}__button--selected`,
        className
    ].filter(Boolean).join(' ');
    if (to) {
        return /*#__PURE__*/ _react.createElement(_reactrouterdom.Link, {
            className: classes,
            id: id,
            onClick: ()=>{
                if (onClick) {
                    onClick();
                }
            },
            to: to
        }, children);
    }
    if (onClick) {
        return /*#__PURE__*/ _react.createElement("button", {
            className: classes,
            disabled: disabled,
            id: id,
            onClick: ()=>{
                if (onClick) {
                    onClick();
                }
            },
            type: "button"
        }, children);
    }
    return /*#__PURE__*/ _react.createElement("div", {
        className: classes,
        id: id
    }, children);
};
