// payload/dist/admin/components/views/ForgotPassword/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _reacttoastify = require("react-toastify");
const _Button = /*#__PURE__*/ _interop_require_default(require("../../elements/Button"));
const _Form = /*#__PURE__*/ _interop_require_default(require("../../forms/Form"));
const _Submit = /*#__PURE__*/ _interop_require_default(require("../../forms/Submit"));
const _Email = /*#__PURE__*/ _interop_require_default(require("../../forms/field-types/Email"));
const _Minimal = /*#__PURE__*/ _interop_require_default(require("../../templates/Minimal"));
const _Auth = require("../../utilities/Auth");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'forgot-password';
const ForgotPassword = ()=>{
    const [hasSubmitted, setHasSubmitted] = (0, _react.useState)(false);
    const { user } = (0, _Auth.useAuth)();
    const { t } = (0, _reacti18next.useTranslation)('authentication');
    const { admin: { user: userSlug }, routes: { admin, api }, serverURL } = (0, _Config.useConfig)();
    const handleResponse = (res)=>{
        res.json().then(()=>{
            setHasSubmitted(true);
        }, ()=>{
            _reacttoastify.toast.error(t('emailNotValid'));
        });
    };
    if (user) {
        return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
            className: baseClass
        }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
            description: t('forgotPassword'),
            keywords: t('forgotPassword'),
            title: t('forgotPassword')
        }), /*#__PURE__*/ _react.default.createElement("h1", null, t('alreadyLoggedIn')), /*#__PURE__*/ _react.default.createElement("p", null, /*#__PURE__*/ _react.default.createElement(_reacti18next.Trans, {
            i18nKey: "loggedInChangePassword",
            t: t
        }, /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
            to: `${admin}/account`
        }, "account"))), /*#__PURE__*/ _react.default.createElement("br", null), /*#__PURE__*/ _react.default.createElement(_Button.default, {
            buttonStyle: "secondary",
            el: "link",
            to: admin
        }, t('general:backToDashboard')));
    }
    if (hasSubmitted) {
        return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
            className: baseClass
        }, /*#__PURE__*/ _react.default.createElement("h1", null, t('emailSent')), /*#__PURE__*/ _react.default.createElement("p", null, t('checkYourEmailForPasswordReset')));
    }
    return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement(_Form.default, {
        action: `${serverURL}${api}/${userSlug}/forgot-password`,
        handleResponse: handleResponse,
        method: "post"
    }, /*#__PURE__*/ _react.default.createElement("h1", null, t('forgotPassword')), /*#__PURE__*/ _react.default.createElement("p", null, t('forgotPasswordEmailInstructions')), /*#__PURE__*/ _react.default.createElement(_Email.default, {
        admin: {
            autoComplete: 'email'
        },
        label: t('general:emailAddress'),
        name: "email",
        required: true
    }), /*#__PURE__*/ _react.default.createElement(_Submit.default, null, t('general:submit'))), /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
        to: `${admin}/login`
    }, t('backToLogin')));
};
const _default = ForgotPassword;
