// payload/dist/admin/components/forms/field-types/RadioGroup/Input.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _types = require("../../../../../fields/config/types");
const _Error = /*#__PURE__*/ _interop_require_default(require("../../Error"));
const _FieldDescription = /*#__PURE__*/ _interop_require_default(require("../../FieldDescription"));
const _Label = /*#__PURE__*/ _interop_require_default(require("../../Label"));
const _shared = require("../shared");
const _RadioInput = /*#__PURE__*/ _interop_require_default(require("./RadioInput"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'radio-group';
const RadioGroupInput = (props)=>{
    const { name, Error, Label, className, description, errorMessage, label, layout = 'horizontal', onChange, options, path: pathFromProps, readOnly, required, showError, style, value, width } = props;
    const ErrorComp = Error || _Error.default;
    const LabelComp = Label || _Label.default;
    const path = pathFromProps || name;
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            _shared.fieldBaseClass,
            baseClass,
            className,
            `${baseClass}--layout-${layout}`,
            showError && 'error',
            readOnly && `${baseClass}--read-only`
        ].filter(Boolean).join(' '),
        style: {
            ...style,
            width
        }
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__error-wrap`
    }, /*#__PURE__*/ _react.default.createElement(ErrorComp, {
        message: errorMessage,
        showError: showError
    })), /*#__PURE__*/ _react.default.createElement(LabelComp, {
        htmlFor: `field-${path}`,
        label: label,
        required: required
    }), /*#__PURE__*/ _react.default.createElement("ul", {
        className: `${baseClass}--group`,
        id: `field-${path.replace(/\./g, '__')}`
    }, options.map((option)=>{
        let optionValue = '';
        if ((0, _types.optionIsObject)(option)) {
            optionValue = option.value;
        } else {
            optionValue = option;
        }
        const isSelected = String(optionValue) === String(value);
        return /*#__PURE__*/ _react.default.createElement("li", {
            key: `${path} - ${optionValue}`
        }, /*#__PURE__*/ _react.default.createElement(_RadioInput.default, {
            isSelected: isSelected,
            onChange: readOnly ? undefined : onChange,
            option: (0, _types.optionIsObject)(option) ? option : {
                label: option,
                value: option
            },
            path: path,
            readOnly: readOnly
        }));
    })), /*#__PURE__*/ _react.default.createElement(_FieldDescription.default, {
        description: description,
        path: path,
        value: value
    }));
};
const _default = RadioGroupInput;
