// payload/dist/admin/components/views/Versions/cells/AutosaveCell.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    AutosaveCell: function() {
        return AutosaveCell;
    },
    renderPill: function() {
        return renderPill;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _ = require("../../../");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const renderPill = (data, latestVersion, currentLabel, previousLabel, pillStyle)=>{
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, data?.id === latestVersion ? /*#__PURE__*/ _react.default.createElement(_.Pill, {
        pillStyle: pillStyle
    }, currentLabel) : /*#__PURE__*/ _react.default.createElement(_.Pill, null, previousLabel), "  ");
};
const AutosaveCell = ({ latestDraftVersion, latestPublishedVersion, rowData })=>{
    const { t } = (0, _reacti18next.useTranslation)();
    const status = rowData?.version._status;
    const versionInfo = {
        draft: {
            currentLabel: t('version:currentDraft'),
            latestVersion: latestDraftVersion,
            pillStyle: undefined,
            previousLabel: t('version:draft')
        },
        published: {
            currentLabel: t('version:currentPublishedVersion'),
            latestVersion: latestPublishedVersion,
            pillStyle: 'success',
            previousLabel: t('version:previouslyPublished')
        }
    };
    const { currentLabel, latestVersion, pillStyle, previousLabel } = versionInfo[status] || {};
    return /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, rowData?.autosave && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_.Pill, null, t('version:autosave')), "  "), status && renderPill(rowData, latestVersion, currentLabel, previousLabel, pillStyle));
};
