// payload/dist/admin/components/modals/LeaveWithoutSaving/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "LeaveWithoutSaving", {
    enumerable: true,
    get: function() {
        return LeaveWithoutSaving;
    }
});
const _modal = require("@faceless-ui/modal");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouternavigationprompt = /*#__PURE__*/ _interop_require_default(require("react-router-navigation-prompt"));
const _Button = /*#__PURE__*/ _interop_require_default(require("../../elements/Button"));
const _context = require("../../forms/Form/context");
const _Auth = require("../../utilities/Auth");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const modalSlug = 'leave-without-saving';
const baseClass = 'leave-without-saving';
const Component = ({ isActive, onCancel, onConfirm })=>{
    const { closeModal, openModal, modalState } = (0, _modal.useModal)();
    const { t } = (0, _reacti18next.useTranslation)('general');
    // Manually check for modal state as 'esc' key will not trigger the nav inactivity
    (0, _react.useEffect)(()=>{
        if (!modalState?.[modalSlug]?.isOpen && isActive) {
            onCancel();
        }
    }, [
        modalState
    ]);
    (0, _react.useEffect)(()=>{
        if (isActive) openModal(modalSlug);
        else closeModal(modalSlug);
    }, [
        isActive,
        openModal,
        closeModal
    ]);
    return /*#__PURE__*/ _react.default.createElement(_modal.Modal, {
        className: baseClass,
        onClose: onCancel,
        slug: modalSlug
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrapper`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__content`
    }, /*#__PURE__*/ _react.default.createElement("h1", null, t('leaveWithoutSaving')), /*#__PURE__*/ _react.default.createElement("p", null, t('changesNotSaved'))), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__controls`
    }, /*#__PURE__*/ _react.default.createElement(_Button.default, {
        buttonStyle: "secondary",
        onClick: onCancel
    }, t('stayOnThisPage')), /*#__PURE__*/ _react.default.createElement(_Button.default, {
        onClick: onConfirm
    }, t('leaveAnyway')))));
};
const LeaveWithoutSaving = ()=>{
    const modified = (0, _context.useFormModified)();
    const { user } = (0, _Auth.useAuth)();
    return /*#__PURE__*/ _react.default.createElement(_reactrouternavigationprompt.default, {
        renderIfNotActive: true,
        when: Boolean(modified && user)
    }, ({ isActive, onCancel, onConfirm })=>/*#__PURE__*/ _react.default.createElement(Component, {
            isActive: isActive,
            onCancel: onCancel,
            onConfirm: onConfirm
        }));
};
