// payload/dist/admin/components/views/LivePreview/Toolbar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "LivePreviewToolbar", {
    enumerable: true,
    get: function() {
        return LivePreviewToolbar;
    }
});
const _core = require("@dnd-kit/core");
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _Drag = /*#__PURE__*/ _interop_require_default(require("../../../icons/Drag"));
const _context = require("../Context/context");
const _Controls = require("./Controls");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'live-preview-toolbar';
const DraggableToolbar = (props)=>{
    const { toolbarPosition } = (0, _context.useLivePreviewContext)();
    const { attributes, listeners, setNodeRef, transform } = (0, _core.useDraggable)({
        id: 'live-preview-toolbar'
    });
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            `${baseClass}--draggable`
        ].join(' '),
        style: {
            left: `${toolbarPosition.x}px`,
            top: `${toolbarPosition.y}px`,
            ...transform ? {
                transform: transform ? `translate3d(${transform?.x || 0}px, ${transform?.y || 0}px, 0)` : undefined
            } : {}
        }
    }, /*#__PURE__*/ _react.default.createElement("button", {
        ...listeners,
        ...attributes,
        className: `${baseClass}__drag-handle`,
        ref: setNodeRef,
        type: "button"
    }, /*#__PURE__*/ _react.default.createElement(_Drag.default, null)), /*#__PURE__*/ _react.default.createElement(_Controls.ToolbarControls, props));
};
const StaticToolbar = (props)=>{
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            `${baseClass}--static`
        ].join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_Controls.ToolbarControls, props));
};
const LivePreviewToolbar = (props)=>{
    const { draggable } = props;
    if (draggable) {
        return /*#__PURE__*/ _react.default.createElement(DraggableToolbar, props);
    }
    return /*#__PURE__*/ _react.default.createElement(StaticToolbar, props);
};
