// payload/dist/globals/requestHandlers/findVersions.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findVersionsHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _isNumber = require("../../utilities/isNumber");
const _findVersions = /*#__PURE__*/ _interop_require_default(require("../operations/findVersions"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function findVersionsHandler(global) {
    return async function handler(req, res, next) {
        try {
            let page;
            if (typeof req.query.page === 'string') {
                const parsedPage = parseInt(req.query.page, 10);
                if (!Number.isNaN(parsedPage)) {
                    page = parsedPage;
                }
            }
            const options = {
                depth: (0, _isNumber.isNumber)(req.query.depth) ? Number(req.query.depth) : undefined,
                globalConfig: global,
                limit: (0, _isNumber.isNumber)(req.query.limit) ? Number(req.query.limit) : undefined,
                page,
                req,
                sort: req.query.sort,
                where: req.query.where
            };
            const result = await (0, _findVersions.default)(options);
            return res.status(_httpstatus.default.OK).json(result);
        } catch (error) {
            return next(error);
        }
    };
}
