// payload/database.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    BaseDatabaseAdapter: function() {
        return _types.BaseDatabaseAdapter;
    },
    BeginTransaction: function() {
        return _types.BeginTransaction;
    },
    CommitTransaction: function() {
        return _types.CommitTransaction;
    },
    Connect: function() {
        return _types.Connect;
    },
    Count: function() {
        return _types.Count;
    },
    CountArgs: function() {
        return _types.CountArgs;
    },
    Create: function() {
        return _types.Create;
    },
    CreateArgs: function() {
        return _types.CreateArgs;
    },
    CreateGlobal: function() {
        return _types.CreateGlobal;
    },
    CreateGlobalArgs: function() {
        return _types.CreateGlobalArgs;
    },
    CreateGlobalVersion: function() {
        return _types.CreateGlobalVersion;
    },
    CreateGlobalVersionArgs: function() {
        return _types.CreateGlobalVersionArgs;
    },
    CreateMigration: function() {
        return _types.CreateMigration;
    },
    CreateVersion: function() {
        return _types.CreateVersion;
    },
    CreateVersionArgs: function() {
        return _types.CreateVersionArgs;
    },
    DBIdentifierName: function() {
        return _types.DBIdentifierName;
    },
    DeleteMany: function() {
        return _types.DeleteMany;
    },
    DeleteManyArgs: function() {
        return _types.DeleteManyArgs;
    },
    DeleteOne: function() {
        return _types.DeleteOne;
    },
    DeleteOneArgs: function() {
        return _types.DeleteOneArgs;
    },
    DeleteVersions: function() {
        return _types.DeleteVersions;
    },
    DeleteVersionsArgs: function() {
        return _types.DeleteVersionsArgs;
    },
    Destroy: function() {
        return _types.Destroy;
    },
    EntityPolicies: function() {
        return _types1.EntityPolicies;
    },
    Find: function() {
        return _types.Find;
    },
    FindArgs: function() {
        return _types.FindArgs;
    },
    FindGlobal: function() {
        return _types.FindGlobal;
    },
    FindGlobalArgs: function() {
        return _types.FindGlobalArgs;
    },
    FindGlobalVersions: function() {
        return _types.FindGlobalVersions;
    },
    FindGlobalVersionsArgs: function() {
        return _types.FindGlobalVersionsArgs;
    },
    FindOne: function() {
        return _types.FindOne;
    },
    FindOneArgs: function() {
        return _types.FindOneArgs;
    },
    FindVersions: function() {
        return _types.FindVersions;
    },
    FindVersionsArgs: function() {
        return _types.FindVersionsArgs;
    },
    Init: function() {
        return _types.Init;
    },
    Migration: function() {
        return _types.Migration;
    },
    MigrationData: function() {
        return _types.MigrationData;
    },
    PaginatedDocs: function() {
        return _types.PaginatedDocs;
    },
    PathToQuery: function() {
        return _types1.PathToQuery;
    },
    QueryDrafts: function() {
        return _types.QueryDrafts;
    },
    QueryDraftsArgs: function() {
        return _types.QueryDraftsArgs;
    },
    RollbackTransaction: function() {
        return _types.RollbackTransaction;
    },
    Transaction: function() {
        return _types.Transaction;
    },
    TypeWithVersion: function() {
        return _types.TypeWithVersion;
    },
    UpdateGlobal: function() {
        return _types.UpdateGlobal;
    },
    UpdateGlobalArgs: function() {
        return _types.UpdateGlobalArgs;
    },
    UpdateGlobalVersion: function() {
        return _types.UpdateGlobalVersion;
    },
    UpdateGlobalVersionArgs: function() {
        return _types.UpdateGlobalVersionArgs;
    },
    UpdateOne: function() {
        return _types.UpdateOne;
    },
    UpdateOneArgs: function() {
        return _types.UpdateOneArgs;
    },
    UpdateVersion: function() {
        return _types.UpdateVersion;
    },
    UpdateVersionArgs: function() {
        return _types.UpdateVersionArgs;
    },
    combineQueries: function() {
        return _combineQueries.combineQueries;
    },
    createDatabaseAdapter: function() {
        return _createDatabaseAdapter.createDatabaseAdapter;
    },
    createMigration: function() {
        return _createMigration.createMigration;
    },
    flattenWhereToOperators: function() {
        return _flattenWhereToOperators.default;
    },
    getLocalizedPaths: function() {
        return _getLocalizedPaths.getLocalizedPaths;
    },
    getMigrations: function() {
        return _getMigrations.getMigrations;
    },
    migrate: function() {
        return _migrate.migrate;
    },
    migrateDown: function() {
        return _migrateDown.migrateDown;
    },
    migrateRefresh: function() {
        return _migrateRefresh.migrateRefresh;
    },
    migrateReset: function() {
        return _migrateReset.migrateReset;
    },
    migrateStatus: function() {
        return _migrateStatus.migrateStatus;
    },
    migrationTemplate: function() {
        return _migrationTemplate.migrationTemplate;
    },
    migrationsCollection: function() {
        return _migrationsCollection.migrationsCollection;
    },
    readMigrationFiles: function() {
        return _readMigrationFiles.readMigrationFiles;
    },
    validateQueryPaths: function() {
        return _validateQueryPaths.validateQueryPaths;
    },
    validateSearchParam: function() {
        return _validateSearchParams.validateSearchParam;
    }
});
const _types = require("./dist/database/types");
const _types1 = _export_star(require("./dist/database/queryValidation/types"), exports);
const _combineQueries = require("./dist/database/combineQueries");
const _createDatabaseAdapter = require("./dist/database/createDatabaseAdapter");
const _flattenWhereToOperators = /*#__PURE__*/ _interop_require_default(require("./dist/database/flattenWhereToOperators"));
const _getLocalizedPaths = require("./dist/database/getLocalizedPaths");
const _createMigration = require("./dist/database/migrations/createMigration");
const _getMigrations = require("./dist/database/migrations/getMigrations");
const _migrate = require("./dist/database/migrations/migrate");
const _migrateDown = require("./dist/database/migrations/migrateDown");
const _migrateRefresh = require("./dist/database/migrations/migrateRefresh");
const _migrateReset = require("./dist/database/migrations/migrateReset");
const _migrateStatus = require("./dist/database/migrations/migrateStatus");
const _migrationTemplate = require("./dist/database/migrations/migrationTemplate");
const _migrationsCollection = require("./dist/database/migrations/migrationsCollection");
const _readMigrationFiles = require("./dist/database/migrations/readMigrationFiles");
const _validateQueryPaths = require("./dist/database/queryValidation/validateQueryPaths");
const _validateSearchParams = require("./dist/database/queryValidation/validateSearchParams");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
