// payload/types.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    FileData: function() {
        return _types.FileData;
    },
    ImageSize: function() {
        return _types.ImageSize;
    },
    IncomingUploadType: function() {
        return _types.IncomingUploadType;
    },
    fieldAffectsData: function() {
        return _types1.fieldAffectsData;
    },
    fieldHasMaxDepth: function() {
        return _types1.fieldHasMaxDepth;
    },
    fieldHasSubFields: function() {
        return _types1.fieldHasSubFields;
    },
    fieldIsArrayType: function() {
        return _types1.fieldIsArrayType;
    },
    fieldIsBlockType: function() {
        return _types1.fieldIsBlockType;
    },
    fieldIsGroupType: function() {
        return _types1.fieldIsGroupType;
    },
    fieldIsLocalized: function() {
        return _types1.fieldIsLocalized;
    },
    fieldIsPresentationalOnly: function() {
        return _types1.fieldIsPresentationalOnly;
    },
    fieldSupportsMany: function() {
        return _types1.fieldSupportsMany;
    },
    optionIsObject: function() {
        return _types1.optionIsObject;
    },
    optionIsValue: function() {
        return _types1.optionIsValue;
    },
    optionsAreObjects: function() {
        return _types1.optionsAreObjects;
    },
    tabHasName: function() {
        return _types1.tabHasName;
    },
    validOperators: function() {
        return _constants.validOperators;
    },
    valueIsValueWithRelation: function() {
        return _types1.valueIsValueWithRelation;
    }
});
_export_star(require("./dist/types"), exports);
const _types = require("./dist/uploads/types");
const _types1 = require("./dist/fields/config/types");
const _constants = require("./dist/types/constants");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
