// payload/dist/admin/components/elements/DocumentHeader/Tabs/Tab/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DocumentTab", {
    enumerable: true,
    get: function() {
        return DocumentTab;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _Config = require("../../../../utilities/Config");
const _DocumentInfo = require("../../../../utilities/DocumentInfo");
require("./index.scss");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'doc-tab';
const DocumentTab = (props)=>{
    const { id, apiURL, collection, condition, global, href: tabHref, isActive: checkIsActive, label, newTab, pillLabel } = props;
    const { t } = (0, _reacti18next.useTranslation)('general');
    const location = (0, _reactrouterdom.useLocation)();
    const config = (0, _Config.useConfig)();
    const documentInfo = (0, _DocumentInfo.useDocumentInfo)();
    const match = (0, _reactrouterdom.useRouteMatch)();
    const { routes } = config;
    const { versions } = documentInfo;
    let href = `${match.url}${typeof tabHref === 'string' ? tabHref : ''}`;
    if (typeof tabHref === 'function') {
        href = tabHref({
            id,
            apiURL,
            collection,
            global,
            match,
            routes
        });
    }
    const isActive = typeof checkIsActive === 'function' ? checkIsActive({
        href,
        location,
        match
    }) : typeof checkIsActive === 'boolean' ? checkIsActive : location.pathname.startsWith(href);
    if (!condition || condition && Boolean(condition({
        collection,
        config,
        documentInfo,
        global
    }))) {
        const labelToRender = typeof label === 'function' ? label({
            t
        }) : label;
        const pillToRender = typeof pillLabel === 'function' ? pillLabel({
            versions
        }) : pillLabel;
        return /*#__PURE__*/ _react.default.createElement("li", {
            className: [
                baseClass,
                isActive && `${baseClass}--active`
            ].filter(Boolean).join(' ')
        }, /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
            className: `${baseClass}__link`,
            to: href,
            ...newTab && {
                rel: 'noopener noreferrer',
                target: '_blank'
            },
            tabIndex: isActive ? -1 : 0
        }, /*#__PURE__*/ _react.default.createElement("span", {
            className: `${baseClass}__label`
        }, labelToRender, pillToRender && /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, " ", /*#__PURE__*/ _react.default.createElement("span", {
            className: `${baseClass}__count`
        }, pillToRender)))));
    }
    return null;
};
