// payload/dist/admin/components/views/NotFound/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _Button = /*#__PURE__*/ _interop_require_default(require("../../elements/Button"));
const _Gutter = require("../../elements/Gutter");
const _StepNav = require("../../elements/StepNav");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'not-found';
const NotFound = (props)=>{
    const { marginTop } = props;
    const { setStepNav } = (0, _StepNav.useStepNav)();
    const { routes: { admin } } = (0, _Config.useConfig)();
    const { t } = (0, _reacti18next.useTranslation)('general');
    (0, _react.useEffect)(()=>{
        setStepNav([
            {
                label: t('notFound')
            }
        ]);
    }, [
        setStepNav,
        t
    ]);
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            marginTop && `${baseClass}--margin-top-${marginTop}`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: t('pageNotFound'),
        keywords: `404 ${t('notFound')}`,
        title: t('notFound')
    }), /*#__PURE__*/ _react.default.createElement(_Gutter.Gutter, {
        className: `${baseClass}__wrap`
    }, /*#__PURE__*/ _react.default.createElement("h1", null, t('nothingFound')), /*#__PURE__*/ _react.default.createElement("p", null, t('sorryNotFound')), /*#__PURE__*/ _react.default.createElement(_Button.default, {
        className: `${baseClass}__button`,
        el: "link",
        to: `${admin}`
    }, t('backToDashboard'))));
};
const _default = NotFound;
