// payload/dist/graphql/registerSchema.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return registerGraphQLSchema;
    }
});
const _graphql = /*#__PURE__*/ _interop_require_wildcard(require("graphql"));
const _graphqlquerycomplexity = /*#__PURE__*/ _interop_require_wildcard(require("graphql-query-complexity"));
const _access = /*#__PURE__*/ _interop_require_default(require("../auth/graphql/resolvers/access"));
const _init = /*#__PURE__*/ _interop_require_default(require("../collections/graphql/init"));
const _init1 = /*#__PURE__*/ _interop_require_default(require("../globals/graphql/init"));
const _buildFallbackLocaleInputType = /*#__PURE__*/ _interop_require_default(require("./schema/buildFallbackLocaleInputType"));
const _buildLocaleInputType = /*#__PURE__*/ _interop_require_default(require("./schema/buildLocaleInputType"));
const _buildPoliciesType = /*#__PURE__*/ _interop_require_default(require("./schema/buildPoliciesType"));
const _wrapCustomResolver = require("./utilities/wrapCustomResolver");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function registerGraphQLSchema(payload) {
    payload.types = {
        arrayTypes: {},
        blockInputTypes: {},
        blockTypes: {},
        groupTypes: {}
    };
    if (payload.config.localization) {
        payload.types.localeInputType = (0, _buildLocaleInputType.default)(payload.config.localization);
        payload.types.fallbackLocaleInputType = (0, _buildFallbackLocaleInputType.default)(payload.config.localization);
    }
    payload.Query = {
        name: 'Query',
        fields: {}
    };
    payload.Mutation = {
        name: 'Mutation',
        fields: {}
    };
    (0, _init.default)(payload);
    (0, _init1.default)(payload);
    payload.Query.fields.Access = {
        resolve: (0, _access.default)(payload),
        type: (0, _buildPoliciesType.default)(payload)
    };
    if (typeof payload.config.graphQL.queries === 'function') {
        const customQueries = payload.config.graphQL.queries(_graphql, payload);
        payload.Query = {
            ...payload.Query,
            fields: {
                ...payload.Query.fields,
                ...(0, _wrapCustomResolver.wrapCustomFields)(customQueries || {})
            }
        };
    }
    if (typeof payload.config.graphQL.mutations === 'function') {
        const customMutations = payload.config.graphQL.mutations(_graphql, payload);
        payload.Mutation = {
            ...payload.Mutation,
            fields: {
                ...payload.Mutation.fields,
                ...(0, _wrapCustomResolver.wrapCustomFields)(customMutations || {})
            }
        };
    }
    const query = new _graphql.GraphQLObjectType(payload.Query);
    const mutation = new _graphql.GraphQLObjectType(payload.Mutation);
    const schema = {
        mutation,
        query
    };
    payload.schema = new _graphql.GraphQLSchema(schema);
    payload.validationRules = ({ variableValues })=>[
            (0, _graphqlquerycomplexity.default)({
                estimators: [
                    (0, _graphqlquerycomplexity.fieldExtensionsEstimator)(),
                    (0, _graphqlquerycomplexity.simpleEstimator)({
                        defaultComplexity: 1
                    })
                ],
                maximumComplexity: payload.config.graphQL.maxComplexity,
                variables: variableValues
            })
        ];
}
