// payload/dist/admin/components/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Banner: function() {
        return _Banner.default;
    },
    Button: function() {
        return _Button.default;
    },
    Check: function() {
        return _Check.default;
    },
    Chevron: function() {
        return _Chevron.default;
    },
    Menu: function() {
        return _Menu.default;
    },
    MinimalTemplate: function() {
        return _Minimal.default;
    },
    Pill: function() {
        return _Pill.default;
    },
    Popup: function() {
        return _Popup.default;
    },
    Search: function() {
        return _Search.default;
    },
    X: function() {
        return _X.default;
    }
});
const _Banner = /*#__PURE__*/ _interop_require_default(require("./elements/Banner"));
const _Button = /*#__PURE__*/ _interop_require_default(require("./elements/Button"));
const _Pill = /*#__PURE__*/ _interop_require_default(require("./elements/Pill"));
const _Popup = /*#__PURE__*/ _interop_require_default(require("./elements/Popup"));
const _Check = /*#__PURE__*/ _interop_require_default(require("./icons/Check"));
const _Chevron = /*#__PURE__*/ _interop_require_default(require("./icons/Chevron"));
const _Menu = /*#__PURE__*/ _interop_require_default(require("./icons/Menu"));
const _Search = /*#__PURE__*/ _interop_require_default(require("./icons/Search"));
const _X = /*#__PURE__*/ _interop_require_default(require("./icons/X"));
const _Minimal = /*#__PURE__*/ _interop_require_default(require("./templates/Minimal"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
