// payload/dist/config/shared/componentSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    componentSchema: function() {
        return componentSchema;
    },
    customViewSchema: function() {
        return customViewSchema;
    },
    documentTabSchema: function() {
        return documentTabSchema;
    },
    livePreviewSchema: function() {
        return livePreviewSchema;
    }
});
const _joi = /*#__PURE__*/ _interop_require_default(require("joi"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const componentSchema = _joi.default.alternatives().try(_joi.default.object().unknown(), _joi.default.func());
const documentTabSchema = {
    condition: _joi.default.func(),
    href: _joi.default.alternatives().try(_joi.default.string(), _joi.default.func()).required(),
    isActive: _joi.default.alternatives().try(_joi.default.func(), _joi.default.boolean()),
    label: _joi.default.alternatives().try(_joi.default.string(), _joi.default.func()).required(),
    newTab: _joi.default.boolean(),
    pillLabel: _joi.default.alternatives().try(_joi.default.string(), _joi.default.func())
};
const customViewSchema = _joi.default.object({
    Component: componentSchema,
    Tab: _joi.default.alternatives().try(documentTabSchema, componentSchema),
    actions: _joi.default.array().items(componentSchema),
    path: _joi.default.string()
});
const livePreviewSchema = {
    breakpoints: _joi.default.array().items(_joi.default.object({
        name: _joi.default.string(),
        height: _joi.default.alternatives().try(_joi.default.number(), _joi.default.string()),
        label: _joi.default.string(),
        width: _joi.default.alternatives().try(_joi.default.number(), _joi.default.string())
    })),
    url: _joi.default.alternatives().try(_joi.default.string(), _joi.default.func())
};
