// payload/dist/versions/baseFields.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    default: function() {
        return _default;
    },
    statuses: function() {
        return statuses;
    }
});
const _extractTranslations = require("../translations/extractTranslations");
const labels = (0, _extractTranslations.extractTranslations)([
    'version:draft',
    'version:published',
    'version:status'
]);
const statuses = [
    {
        label: labels['version:draft'],
        value: 'draft'
    },
    {
        label: labels['version:published'],
        value: 'published'
    }
];
const baseVersionFields = [
    {
        name: '_status',
        type: 'select',
        admin: {
            components: {
                Field: ()=>null
            },
            disableBulkEdit: true
        },
        defaultValue: 'draft',
        index: true,
        label: labels['version:status'],
        options: statuses
    }
];
const _default = baseVersionFields;
