// payload/dist/versions/buildGlobalFields.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "buildVersionGlobalFields", {
    enumerable: true,
    get: function() {
        return buildVersionGlobalFields;
    }
});
const buildVersionGlobalFields = (global)=>{
    const fields = [
        {
            name: 'version',
            fields: global.fields,
            type: 'group'
        },
        {
            name: 'createdAt',
            admin: {
                disabled: true
            },
            type: 'date'
        },
        {
            name: 'updatedAt',
            admin: {
                disabled: true
            },
            type: 'date'
        }
    ];
    if (global?.versions?.drafts) {
        fields.push({
            name: 'latest',
            admin: {
                disabled: true
            },
            index: true,
            type: 'checkbox'
        });
    }
    if (global?.versions?.drafts && global?.versions?.drafts?.autosave) {
        fields.push({
            name: 'autosave',
            index: true,
            type: 'checkbox'
        });
    }
    return fields;
};
