// payload/dist/versions/getLatestGlobalVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getLatestGlobalVersion", {
    enumerable: true,
    get: function() {
        return getLatestGlobalVersion;
    }
});
const _types = require("../types");
const getLatestGlobalVersion = async ({ config, locale, payload, req, slug, where })=>{
    let latestVersion;
    if (config.versions?.drafts) {
        // eslint-disable-next-line prefer-destructuring
        latestVersion = (await payload.db.findGlobalVersions({
            global: slug,
            limit: 1,
            locale,
            req,
            sort: '-updatedAt'
        })).docs[0];
    }
    const global = await payload.db.findGlobal({
        locale,
        req,
        slug,
        where
    });
    const globalExists = Boolean(global);
    if (!latestVersion || (0, _types.docHasTimestamps)(global) && latestVersion.updatedAt < global.updatedAt) {
        return {
            global,
            globalExists
        };
    }
    return {
        global: {
            ...latestVersion.version,
            createdAt: latestVersion.createdAt,
            updatedAt: latestVersion.updatedAt
        },
        globalExists
    };
};
