// payload/dist/admin/components/elements/ReactSelect/Control/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Control", {
    enumerable: true,
    get: function() {
        return Control;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reactselect = require("react-select");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const Control = (props)=>{
    const { children, innerProps, // @ts-expect-error // TODO Fix this - moduleResolution 16 breaks our declare module
    selectProps: { customProps: { disableKeyDown, disableMouseDown } = {} } = {} } = props;
    return /*#__PURE__*/ _react.default.createElement(_reactselect.components.Control, {
        ...props,
        innerProps: {
            ...innerProps,
            // @ts-ignore
            onKeyDown: (e)=>{
                if (disableKeyDown) {
                    e.stopPropagation();
                    // Create event for keydown listeners which specifically want to bypass this stopPropagation
                    const bypassEvent = new CustomEvent('bypassKeyDown', {
                        detail: e
                    });
                    document.dispatchEvent(bypassEvent);
                }
            },
            // react-select has this typed incorrectly so we disable the linting rule
            // we need to prevent react-select from hijacking the 'onKeyDown' event while modals are open (i.e. the 'Relationship' field component)
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            onMouseDown: (e)=>{
                // we need to prevent react-select from hijacking the 'onMouseDown' event while modals are open (i.e. the 'Relationship' field component)
                if (!disableMouseDown) {
                    innerProps.onMouseDown(e);
                }
            }
        }
    }, children);
};
