// payload/dist/admin/components/forms/field-types/Blocks/RowActions.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "RowActions", {
    enumerable: true,
    get: function() {
        return RowActions;
    }
});
const _modal = require("@faceless-ui/modal");
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _ArrayAction = require("../../../elements/ArrayAction");
const _useDrawerSlug = require("../../../elements/Drawer/useDrawerSlug");
const _BlocksDrawer = require("./BlocksDrawer");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const RowActions = (props)=>{
    const { addRow, blockType, blocks, duplicateRow, hasMaxRows, isSortable, labels, moveRow, removeRow, rowCount, rowIndex } = props;
    const { closeModal, openModal } = (0, _modal.useModal)();
    const drawerSlug = (0, _useDrawerSlug.useDrawerSlug)('blocks-drawer');
    const [indexToAdd, setIndexToAdd] = _react.default.useState(null);
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_BlocksDrawer.BlocksDrawer, {
        addRow: (_, rowBlockType)=>{
            if (typeof addRow === 'function') {
                addRow(indexToAdd, rowBlockType);
            }
            closeModal(drawerSlug);
        },
        addRowIndex: rowIndex,
        blocks: blocks,
        drawerSlug: drawerSlug,
        labels: labels
    }), /*#__PURE__*/ _react.default.createElement(_ArrayAction.ArrayAction, {
        addRow: (index)=>{
            setIndexToAdd(index);
            openModal(drawerSlug);
        },
        duplicateRow: ()=>duplicateRow(rowIndex, blockType),
        hasMaxRows: hasMaxRows,
        index: rowIndex,
        isSortable: isSortable,
        moveRow: moveRow,
        removeRow: removeRow,
        rowCount: rowCount
    }));
};
