// payload/dist/admin/components/graphics/Account/Default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DefaultAccountIcon", {
    enumerable: true,
    get: function() {
        return DefaultAccountIcon;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'graphic-account';
const DefaultAccountIcon = (props)=>/*#__PURE__*/ _react.default.createElement("svg", {
        className: [
            baseClass,
            props?.active && `${baseClass}--active`
        ].filter(Boolean).join(' '),
        height: "25",
        viewBox: "0 0 25 25",
        width: "25",
        xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/ _react.default.createElement("circle", {
        className: `${baseClass}__bg`,
        cx: "12.5",
        cy: "12.5",
        r: "11.5"
    }), /*#__PURE__*/ _react.default.createElement("circle", {
        className: `${baseClass}__head`,
        cx: "12.5",
        cy: "10.73",
        r: "3.98"
    }), /*#__PURE__*/ _react.default.createElement("path", {
        className: `${baseClass}__body`,
        d: "M12.5,24a11.44,11.44,0,0,0,7.66-2.94c-.5-2.71-3.73-4.8-7.66-4.8s-7.16,2.09-7.66,4.8A11.44,11.44,0,0,0,12.5,24Z"
    }));
