// payload/dist/admin/components/elements/SortComplex/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _types = require("../../../../fields/config/types");
const _sortableFieldTypes = /*#__PURE__*/ _interop_require_default(require("../../../../fields/sortableFieldTypes"));
const _getTranslation = require("../../../../utilities/getTranslation");
const _SearchParams = require("../../utilities/SearchParams");
const _ReactSelect = /*#__PURE__*/ _interop_require_default(require("../ReactSelect"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'sort-complex';
const SortComplex = (props)=>{
    const { collection, handleChange, modifySearchQuery = true } = props;
    const history = (0, _reactrouterdom.useHistory)();
    const params = (0, _SearchParams.useSearchParams)();
    const { i18n, t } = (0, _reacti18next.useTranslation)('general');
    const [sortOptions, setSortOptions] = (0, _react.useState)();
    const [sortFields] = (0, _react.useState)(()=>collection.fields.reduce((fields, field)=>{
            if ((0, _types.fieldAffectsData)(field) && _sortableFieldTypes.default.indexOf(field.type) > -1) {
                return [
                    ...fields,
                    {
                        label: (0, _getTranslation.getTranslation)(field.label || field.name, i18n),
                        value: field.name
                    }
                ];
            }
            return fields;
        }, []));
    const [sortField, setSortField] = (0, _react.useState)(sortFields[0]);
    const [initialSort] = (0, _react.useState)(()=>({
            label: t('descending'),
            value: '-'
        }));
    const [sortOrder, setSortOrder] = (0, _react.useState)(initialSort);
    (0, _react.useEffect)(()=>{
        if (sortField?.value) {
            const newSortValue = `${sortOrder.value}${sortField.value}`;
            if (handleChange) handleChange(newSortValue);
            if (params.sort !== newSortValue && modifySearchQuery) {
                history.replace({
                    search: _qs.default.stringify({
                        ...params,
                        sort: newSortValue
                    }, {
                        addQueryPrefix: true
                    })
                });
            }
        }
    }, [
        history,
        params,
        sortField,
        sortOrder,
        modifySearchQuery,
        handleChange
    ]);
    (0, _react.useEffect)(()=>{
        setSortOptions([
            {
                label: t('ascending'),
                value: ''
            },
            {
                label: t('descending'),
                value: '-'
            }
        ]);
    }, [
        i18n,
        t
    ]);
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrap`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__select`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__label`
    }, t('columnToSort')), /*#__PURE__*/ _react.default.createElement(_ReactSelect.default, {
        onChange: setSortField,
        options: sortFields,
        value: sortField
    })), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__select`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__label`
    }, t('order')), /*#__PURE__*/ _react.default.createElement(_ReactSelect.default, {
        onChange: (incomingSort)=>{
            setSortOrder(incomingSort || initialSort);
        },
        options: sortOptions,
        value: sortOrder
    })))));
};
const _default = SortComplex;
