// payload/dist/auth/requestHandlers/forgotPassword.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return forgotPasswordHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _forgotPassword = /*#__PURE__*/ _interop_require_default(require("../operations/forgotPassword"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function forgotPasswordHandler(req, res, next) {
    try {
        await (0, _forgotPassword.default)({
            collection: req.collection,
            data: {
                email: req.body.email
            },
            disableEmail: req.body.disableEmail,
            expiration: req.body.expiration,
            req
        });
        return res.status(_httpstatus.default.OK).json({
            message: 'Success'
        });
    } catch (error) {
        return next(error);
    }
}
