// payload/dist/auth/requestHandlers/login.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return loginHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _login = /*#__PURE__*/ _interop_require_default(require("../operations/login"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function loginHandler(req, res, next) {
    try {
        const result = await (0, _login.default)({
            collection: req.collection,
            data: req.body,
            depth: parseInt(String(req.query.depth), 10),
            req,
            res
        });
        res.status(_httpstatus.default.OK).json({
            exp: result.exp,
            message: 'Auth Passed',
            token: result.token,
            user: result.user
        });
    } catch (error) {
        next(error);
    }
}
