// payload/dist/auth/requestHandlers/resetPassword.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _resetPassword = /*#__PURE__*/ _interop_require_default(require("../operations/resetPassword"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function resetPasswordHandler(req, res, next) {
    try {
        const result = await (0, _resetPassword.default)({
            collection: req.collection,
            data: req.body,
            req,
            res
        });
        return res.status(_httpstatus.default.OK).json({
            message: 'Password reset successfully.',
            token: result.token,
            user: result.user
        });
    } catch (error) {
        return next(error);
    }
}
const _default = resetPasswordHandler;
