// payload/dist/admin/components/elements/NavGroup/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reactanimateheight = /*#__PURE__*/ _interop_require_default(require("react-animate-height"));
const _Chevron = /*#__PURE__*/ _interop_require_default(require("../../icons/Chevron"));
const _Preferences = require("../../utilities/Preferences");
const _context = require("../Nav/context");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'nav-group';
const NavGroup = ({ children, label })=>{
    const [collapsed, setCollapsed] = (0, _react.useState)(true);
    const [animate, setAnimate] = (0, _react.useState)(false);
    const { getPreference, setPreference } = (0, _Preferences.usePreferences)();
    const { navOpen } = (0, _context.useNav)();
    const preferencesKey = `collapsed-${label}-groups`;
    (0, _react.useEffect)(()=>{
        if (label) {
            const setCollapsedFromPreferences = async ()=>{
                const preferences = await getPreference(preferencesKey) || [];
                setCollapsed(preferences.indexOf(label) !== -1);
            };
            setCollapsedFromPreferences();
        }
    }, [
        getPreference,
        label,
        preferencesKey
    ]);
    if (label) {
        const toggleCollapsed = async ()=>{
            setAnimate(true);
            let preferences = await getPreference(preferencesKey) || [];
            if (collapsed) {
                preferences = preferences.filter((preference)=>label !== preference);
            } else {
                preferences.push(label);
            }
            setPreference(preferencesKey, preferences);
            setCollapsed(!collapsed);
        };
        return /*#__PURE__*/ _react.default.createElement("div", {
            className: [
                `${baseClass}`,
                `${label}`,
                collapsed && `${baseClass}--collapsed`
            ].filter(Boolean).join(' '),
            id: `nav-group-${label}`
        }, /*#__PURE__*/ _react.default.createElement("button", {
            className: [
                `${baseClass}__toggle`,
                `${baseClass}__toggle--${collapsed ? 'collapsed' : 'open'}`
            ].filter(Boolean).join(' '),
            onClick: toggleCollapsed,
            tabIndex: !navOpen ? -1 : 0,
            type: "button"
        }, /*#__PURE__*/ _react.default.createElement("div", {
            className: `${baseClass}__label`
        }, label), /*#__PURE__*/ _react.default.createElement("div", {
            className: `${baseClass}__indicator`
        }, /*#__PURE__*/ _react.default.createElement(_Chevron.default, {
            className: `${baseClass}__indicator`,
            direction: !collapsed ? 'up' : undefined
        }))), /*#__PURE__*/ _react.default.createElement(_reactanimateheight.default, {
            duration: animate ? 200 : 0,
            height: collapsed ? 0 : 'auto'
        }, /*#__PURE__*/ _react.default.createElement("div", {
            className: `${baseClass}__content`
        }, children)));
    }
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, children);
};
const _default = NavGroup;
