// payload/dist/admin/components/utilities/Meta/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacthelmet = /*#__PURE__*/ _interop_require_default(require("react-helmet"));
const _reacti18next = require("react-i18next");
const _faviconsvg = /*#__PURE__*/ _interop_require_default(require("../../../assets/images/favicon.svg"));
const _ogimagepng = /*#__PURE__*/ _interop_require_default(require("../../../assets/images/og-image.png"));
const _useMountEffect = /*#__PURE__*/ _interop_require_default(require("../../../hooks/useMountEffect"));
const _Config = require("../Config");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const rtlLanguages = [
    'ar',
    'fa',
    'ha',
    'ku',
    'ur',
    'ps',
    'dv',
    'ks',
    'khw',
    'he',
    'yi'
];
const Meta = ({ description, keywords = 'CMS, Admin, Dashboard', // lang = 'en',
meta = [], title })=>{
    const { i18n } = (0, _reacti18next.useTranslation)();
    const currentLanguage = i18n.language;
    const currentDirection = rtlLanguages.includes(currentLanguage) ? 'RTL' : 'LTR';
    const config = (0, _Config.useConfig)();
    const titleSuffix = config.admin.meta?.titleSuffix ?? '- Payload';
    const favicon = config.admin.meta.favicon ?? _faviconsvg.default;
    const ogImage = config.admin.meta.ogImage ?? _ogimagepng.default;
    (0, _useMountEffect.default)(()=>{
        const faviconElement = document.querySelector('link[data-placeholder-favicon]');
        if (faviconElement) {
            faviconElement.remove();
        }
    });
    return /*#__PURE__*/ _react.default.createElement(_reacthelmet.default, {
        htmlAttributes: {
            dir: currentDirection,
            lang: currentLanguage
        },
        link: [
            {
                href: favicon,
                rel: 'icon',
                type: 'image/svg+xml'
            }
        ],
        meta: [
            {
                name: 'description',
                content: description
            },
            {
                name: 'keywords',
                content: keywords
            },
            {
                content: `${title} ${titleSuffix}`,
                property: 'og:title'
            },
            {
                content: ogImage,
                property: 'og:image'
            },
            {
                content: description,
                property: 'og:description'
            },
            {
                content: 'website',
                property: 'og:type'
            },
            {
                name: 'twitter:card',
                content: 'summary'
            },
            {
                name: 'twitter:title',
                content: title
            },
            {
                name: 'twitter:description',
                content: description
            }
        ].concat(meta),
        title: `${title} ${titleSuffix}`
    });
};
const _default = Meta;
