// payload/dist/admin/components/elements/SearchFilter/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _getTranslation = require("../../../../utilities/getTranslation");
const _useDebounce = /*#__PURE__*/ _interop_require_default(require("../../../hooks/useDebounce"));
const _Search = /*#__PURE__*/ _interop_require_default(require("../../icons/Search"));
const _SearchParams = require("../../utilities/SearchParams");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'search-filter';
const SearchFilter = (props)=>{
    const { fieldLabel = 'ID', fieldName = 'id', handleChange, listSearchableFields, modifySearchQuery = true } = props;
    const params = (0, _SearchParams.useSearchParams)();
    const history = (0, _reactrouterdom.useHistory)();
    const { i18n, t } = (0, _reacti18next.useTranslation)('general');
    const [search, setSearch] = (0, _react.useState)(typeof params?.search === 'string' ? params?.search : '');
    const [previousSearch, setPreviousSearch] = (0, _react.useState)('');
    const placeholder = (0, _react.useRef)(t('searchBy', {
        label: (0, _getTranslation.getTranslation)(fieldLabel, i18n)
    }));
    const debouncedSearch = (0, _useDebounce.default)(search, 300);
    (0, _react.useEffect)(()=>{
        if (debouncedSearch !== previousSearch) {
            if (handleChange) handleChange(debouncedSearch);
            if (modifySearchQuery) {
                history.replace({
                    search: _qs.default.stringify({
                        ...params,
                        page: 1,
                        search: debouncedSearch || undefined
                    })
                });
            }
            setPreviousSearch(debouncedSearch);
        }
    }, [
        debouncedSearch,
        previousSearch,
        history,
        fieldName,
        params,
        handleChange,
        modifySearchQuery,
        listSearchableFields
    ]);
    (0, _react.useEffect)(()=>{
        if (listSearchableFields?.length > 0) {
            placeholder.current = listSearchableFields.reduce((prev, curr, i)=>{
                if (i === 0) {
                    return `${t('searchBy', {
                        label: (0, _getTranslation.getTranslation)(curr.label || curr.name, i18n)
                    })}`;
                }
                if (i === listSearchableFields.length - 1) {
                    return `${prev} ${t('or')} ${(0, _getTranslation.getTranslation)(curr.label || curr.name, i18n)}`;
                }
                return `${prev}, ${(0, _getTranslation.getTranslation)(curr.label || curr.name, i18n)}`;
            }, '');
        } else {
            placeholder.current = t('searchBy', {
                label: (0, _getTranslation.getTranslation)(fieldLabel, i18n)
            });
        }
    }, [
        t,
        listSearchableFields,
        i18n,
        fieldLabel
    ]);
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("input", {
        className: `${baseClass}__input`,
        onChange: (e)=>setSearch(e.target.value),
        placeholder: placeholder.current,
        type: "text",
        value: search || ''
    }), /*#__PURE__*/ _react.default.createElement(_Search.default, null));
};
const _default = SearchFilter;
