// payload/dist/admin/components/elements/CopyToClipboard/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _Copy = /*#__PURE__*/ _interop_require_default(require("../../icons/Copy"));
const _Tooltip = /*#__PURE__*/ _interop_require_default(require("../Tooltip"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'copy-to-clipboard';
const CopyToClipboard = ({ defaultMessage, successMessage, value })=>{
    const ref = (0, _react.useRef)(null);
    const [copied, setCopied] = (0, _react.useState)(false);
    const [hovered, setHovered] = (0, _react.useState)(false);
    const { t } = (0, _reacti18next.useTranslation)('general');
    if (value) {
        return /*#__PURE__*/ _react.default.createElement("button", {
            className: baseClass,
            onClick: ()=>{
                if (ref && ref.current) {
                    ref.current.select();
                    ref.current.setSelectionRange(0, value.length + 1);
                    document.execCommand('copy');
                    setCopied(true);
                }
            },
            onMouseEnter: ()=>{
                setHovered(true);
                setCopied(false);
            },
            onMouseLeave: ()=>{
                setHovered(false);
                setCopied(false);
            },
            type: "button"
        }, /*#__PURE__*/ _react.default.createElement(_Copy.default, null), /*#__PURE__*/ _react.default.createElement(_Tooltip.default, {
            delay: copied ? 0 : undefined,
            show: hovered || copied
        }, copied && (successMessage ?? t('copied')), !copied && (defaultMessage ?? t('copy'))), /*#__PURE__*/ _react.default.createElement("textarea", {
            readOnly: true,
            ref: ref,
            tabIndex: -1,
            value: value
        }));
    }
    return null;
};
const _default = CopyToClipboard;
