// payload/dist/exports/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    APIError: function() {
        return _errors.APIError;
    },
    AuthenticationError: function() {
        return _errors.AuthenticationError;
    },
    DuplicateCollection: function() {
        return _errors.DuplicateCollection;
    },
    DuplicateGlobal: function() {
        return _errors.DuplicateGlobal;
    },
    ErrorDeletingFile: function() {
        return _errors.ErrorDeletingFile;
    },
    FileUploadError: function() {
        return _errors.FileUploadError;
    },
    Forbidden: function() {
        return _errors.Forbidden;
    },
    InvalidConfiguration: function() {
        return _errors.InvalidConfiguration;
    },
    InvalidFieldName: function() {
        return _errors.InvalidFieldName;
    },
    InvalidFieldRelationship: function() {
        return _errors.InvalidFieldRelationship;
    },
    LockedAuth: function() {
        return _errors.LockedAuth;
    },
    MissingCollectionLabel: function() {
        return _errors.MissingCollectionLabel;
    },
    MissingFieldInputOptions: function() {
        return _errors.MissingFieldInputOptions;
    },
    MissingFieldType: function() {
        return _errors.MissingFieldType;
    },
    MissingFile: function() {
        return _errors.MissingFile;
    },
    NotFound: function() {
        return _errors.NotFound;
    },
    QueryError: function() {
        return _errors.QueryError;
    },
    ValidationError: function() {
        return _errors.ValidationError;
    }
});
const _errors = require("../errors");
