// payload/dist/admin/components/forms/field-types/JSON/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _validations = require("../../../../../fields/validations");
const _CodeEditor = require("../../../elements/CodeEditor");
const _Error = /*#__PURE__*/ _interop_require_default(require("../../Error"));
const _FieldDescription = /*#__PURE__*/ _interop_require_default(require("../../FieldDescription"));
const _Label = /*#__PURE__*/ _interop_require_default(require("../../Label"));
const _useField = /*#__PURE__*/ _interop_require_default(require("../../useField"));
const _withCondition = /*#__PURE__*/ _interop_require_default(require("../../withCondition"));
const _shared = require("../shared");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'json-field';
const JSONField = (props)=>{
    const { name, admin: { className, components: { Error, Label } = {}, condition, description, editorOptions, readOnly, style, width } = {}, jsonSchema, label, path: pathFromProps, required, validate = _validations.json } = props;
    const ErrorComp = Error || _Error.default;
    const LabelComp = Label || _Label.default;
    const path = pathFromProps || name;
    const [stringValue, setStringValue] = (0, _react.useState)();
    const [jsonError, setJsonError] = (0, _react.useState)();
    const [hasLoadedValue, setHasLoadedValue] = (0, _react.useState)(false);
    const memoizedValidate = (0, _react.useCallback)((value, options)=>{
        return validate(value, {
            ...options,
            jsonError,
            required
        });
    }, [
        validate,
        required,
        jsonError
    ]);
    const { errorMessage, initialValue, setValue, showError, value } = (0, _useField.default)({
        condition,
        path,
        validate: memoizedValidate
    });
    const handleMount = (0, _react.useCallback)((editor, monaco)=>{
        if (!jsonSchema) return;
        const existingSchemas = monaco.languages.json.jsonDefaults.diagnosticsOptions.schemas || [];
        const modelUri = monaco.Uri.parse(jsonSchema.uri);
        const model = monaco.editor.createModel(JSON.stringify(value, null, 2), 'json', modelUri);
        monaco.languages.json.jsonDefaults.setDiagnosticsOptions({
            enableSchemaRequest: true,
            schemas: [
                ...existingSchemas,
                jsonSchema
            ],
            validate: true
        });
        editor.setModel(model);
    }, [
        value,
        jsonSchema
    ]);
    const handleChange = (0, _react.useCallback)((val)=>{
        try {
            if (readOnly) return;
            setStringValue(val);
            setValue(val ? JSON.parse(val) : '');
            setJsonError(undefined);
        } catch (e) {
            setJsonError(e);
        }
    }, [
        readOnly,
        setValue,
        setStringValue
    ]);
    (0, _react.useEffect)(()=>{
        try {
            const hasValue = value && value.toString().length > 0;
            if (hasLoadedValue) {
                setStringValue(hasValue ? JSON.stringify(value, null, 2) : '');
            } else {
                setStringValue(JSON.stringify(hasValue ? value : initialValue, null, 2));
                setHasLoadedValue(true);
            }
        } catch (e) {
            setJsonError(e);
        }
    }, [
        initialValue,
        value,
        hasLoadedValue
    ]);
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            _shared.fieldBaseClass,
            baseClass,
            className,
            showError && 'error',
            readOnly && 'read-only'
        ].filter(Boolean).join(' '),
        style: {
            ...style,
            width
        }
    }, /*#__PURE__*/ _react.default.createElement(ErrorComp, {
        message: errorMessage,
        showError: showError
    }), /*#__PURE__*/ _react.default.createElement(LabelComp, {
        htmlFor: `field-${path}`,
        label: label,
        required: required
    }), /*#__PURE__*/ _react.default.createElement(_CodeEditor.CodeEditor, {
        defaultLanguage: "json",
        onChange: handleChange,
        onMount: handleMount,
        options: editorOptions,
        readOnly: readOnly,
        value: stringValue
    }), /*#__PURE__*/ _react.default.createElement(_FieldDescription.default, {
        description: description,
        path: path,
        value: value
    }));
};
const _default = (0, _withCondition.default)(JSONField);
