// payload/dist/preferences/deleteUserPreferences.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "deleteUserPreferences", {
    enumerable: true,
    get: function() {
        return deleteUserPreferences;
    }
});
const deleteUserPreferences = async ({ collectionConfig, ids, payload, req })=>{
    if (collectionConfig.auth) {
        await payload.db.deleteMany({
            collection: 'payload-preferences',
            req,
            where: {
                and: [
                    {
                        'user.value': {
                            in: ids
                        }
                    },
                    {
                        'user.relationTo': {
                            equals: collectionConfig.slug
                        }
                    }
                ]
            }
        });
    }
    await payload.db.deleteMany({
        collection: 'payload-preferences',
        req,
        where: {
            key: {
                in: ids.map((id)=>`collection-${collectionConfig.slug}-${id}`)
            }
        }
    });
};
