// payload/dist/admin/components/views/LivePreview/Context/context.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    LivePreviewContext: function() {
        return LivePreviewContext;
    },
    useLivePreviewContext: function() {
        return useLivePreviewContext;
    }
});
const _react = require("react");
const LivePreviewContext = (0, _react.createContext)({
    appIsReady: false,
    breakpoint: undefined,
    breakpoints: undefined,
    fieldSchemaJSON: undefined,
    iframeHasLoaded: false,
    iframeRef: undefined,
    isPopupOpen: false,
    measuredDeviceSize: {
        height: 0,
        width: 0
    },
    openPopupWindow: ()=>{},
    popupRef: undefined,
    previewWindowType: 'iframe',
    setAppIsReady: ()=>{},
    setBreakpoint: ()=>{},
    setHeight: ()=>{},
    setIframeHasLoaded: ()=>{},
    setMeasuredDeviceSize: ()=>{},
    setPreviewWindowType: ()=>{},
    setSize: ()=>{},
    setToolbarPosition: ()=>{},
    setWidth: ()=>{},
    setZoom: ()=>{},
    size: {
        height: 0,
        width: 0
    },
    toolbarPosition: {
        x: 0,
        y: 0
    },
    url: undefined,
    zoom: 1
});
const useLivePreviewContext = ()=>(0, _react.useContext)(LivePreviewContext);
