// payload/dist/admin/components/views/Account/Default.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _DocumentControls = require("../../elements/DocumentControls");
const _DocumentFields = require("../../elements/DocumentFields");
const _DocumentHeader = require("../../elements/DocumentHeader");
const _Loading = require("../../elements/Loading");
const _Form = /*#__PURE__*/ _interop_require_default(require("../../forms/Form"));
const _LeaveWithoutSaving = require("../../modals/LeaveWithoutSaving");
const _Auth = require("../../utilities/Auth");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
const _OperationProvider = require("../../utilities/OperationProvider");
const _Auth1 = /*#__PURE__*/ _interop_require_default(require("../collections/Edit/Auth"));
const _Settings = require("./Settings");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'account';
const DefaultAccount = (props)=>{
    const { action, apiURL, collection, data, fieldTypes, hasSavePermission, initialState, isLoading, onSave: onSaveFromProps, permissions } = props;
    const { auth, fields } = collection;
    const { refreshCookieAsync } = (0, _Auth.useAuth)();
    const { t } = (0, _reacti18next.useTranslation)('authentication');
    const onSave = (0, _react.useCallback)(async (json)=>{
        await refreshCookieAsync();
        if (typeof onSaveFromProps === 'function') {
            onSaveFromProps(json);
        }
    }, [
        onSaveFromProps,
        refreshCookieAsync
    ]);
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: t('accountOfCurrentUser'),
        keywords: t('account'),
        title: t('account')
    }), /*#__PURE__*/ _react.default.createElement(_Loading.LoadingOverlayToggle, {
        name: "account",
        show: isLoading,
        type: "withoutNav"
    }), !isLoading && /*#__PURE__*/ _react.default.createElement(_OperationProvider.OperationContext.Provider, {
        value: "update"
    }, /*#__PURE__*/ _react.default.createElement(_Form.default, {
        action: action,
        disabled: !hasSavePermission,
        initialState: initialState,
        method: "patch",
        onSuccess: onSave
    }, !(collection.versions?.drafts && collection.versions?.drafts?.autosave) && /*#__PURE__*/ _react.default.createElement(_LeaveWithoutSaving.LeaveWithoutSaving, null), /*#__PURE__*/ _react.default.createElement(_DocumentHeader.DocumentHeader, {
        apiURL: apiURL,
        collection: collection,
        data: data
    }), /*#__PURE__*/ _react.default.createElement(_DocumentControls.DocumentControls, {
        apiURL: apiURL,
        collection: collection,
        data: data,
        hasSavePermission: hasSavePermission,
        isAccountView: true,
        permissions: permissions
    }), /*#__PURE__*/ _react.default.createElement(_DocumentFields.DocumentFields, {
        AfterFields: /*#__PURE__*/ _react.default.createElement(_Settings.Settings, {
            className: `${baseClass}__settings`
        }),
        BeforeFields: /*#__PURE__*/ _react.default.createElement(_Auth1.default, {
            className: `${baseClass}__auth`,
            collection: collection,
            email: data?.email,
            operation: "update",
            readOnly: !hasSavePermission,
            useAPIKey: auth.useAPIKey
        }),
        fieldTypes: fieldTypes,
        fields: fields,
        hasSavePermission: hasSavePermission,
        permissions: permissions
    }))));
};
const _default = DefaultAccount;
