// payload/dist/admin/components/graphics/Account/Gravatar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _md5 = /*#__PURE__*/ _interop_require_default(require("md5"));
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _Auth = require("../../../utilities/Auth");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const Gravatar = ()=>{
    const { user } = (0, _Auth.useAuth)();
    const hash = (0, _md5.default)(user.email.trim().toLowerCase());
    const query = _qs.default.stringify({
        default: 'mp',
        r: 'g',
        s: 50
    });
    return /*#__PURE__*/ _react.default.createElement("img", {
        alt: "yas",
        className: "gravatar-account",
        height: 25,
        src: `https://www.gravatar.com/avatar/${hash}?${query}`,
        style: {
            borderRadius: '50%'
        },
        width: 25
    });
};
const _default = Gravatar;
