// payload/dist/auth/strategies/apiKey.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _crypto = /*#__PURE__*/ _interop_require_default(require("crypto"));
const _passportheaderapikey = /*#__PURE__*/ _interop_require_default(require("passport-headerapikey"));
const _find = /*#__PURE__*/ _interop_require_default(require("../../collections/operations/find"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const _default = (payload, config)=>{
    const { secret } = payload;
    const opts = {
        header: 'Authorization',
        prefix: `${config.slug} API-Key `
    };
    return new _passportheaderapikey.default(opts, true, async (apiKey, done, req)=>{
        const apiKeyIndex = _crypto.default.createHmac('sha1', secret).update(apiKey).digest('hex');
        try {
            const where = {};
            if (config.auth.verify) {
                where.and = [
                    {
                        // TODO: Search for index
                        apiKeyIndex: {
                            equals: apiKeyIndex
                        }
                    },
                    {
                        _verified: {
                            not_equals: false
                        }
                    }
                ];
            } else {
                where.apiKeyIndex = {
                    equals: apiKeyIndex
                };
            }
            const userQuery = await (0, _find.default)({
                collection: {
                    config
                },
                depth: config.auth.depth,
                overrideAccess: true,
                req: req,
                where
            });
            if (userQuery.docs && userQuery.docs.length > 0) {
                const user = userQuery.docs[0];
                user.collection = config.slug;
                user._strategy = 'api-key';
                done(null, user);
            } else {
                done(null, false);
            }
        } catch (err) {
            done(null, false);
        }
    });
};
