// payload/dist/auth/strategies/jwt.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _passportjwt = /*#__PURE__*/ _interop_require_default(require("passport-jwt"));
const _url = /*#__PURE__*/ _interop_require_default(require("url"));
const _getExtractJWT = /*#__PURE__*/ _interop_require_default(require("../getExtractJWT"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const JwtStrategy = _passportjwt.default.Strategy;
const _default = ({ collections, config, secret })=>{
    const opts = {
        jwtFromRequest: (0, _getExtractJWT.default)(config),
        passReqToCallback: true,
        secretOrKey: secret
    };
    return new JwtStrategy(opts, async (req, token, done)=>{
        if (req.user) {
            done(null, req.user);
        }
        try {
            const collection = collections[token.collection];
            const parsedURL = _url.default.parse(req.originalUrl);
            const isGraphQL = parsedURL.pathname === `/api${req.payload.config.routes.graphQL}`;
            const user = await req.payload.findByID({
                id: token.id,
                collection: token.collection,
                depth: isGraphQL ? 0 : collection.config.auth.depth,
                req
            });
            if (user && (!collection.config.auth.verify || user._verified)) {
                user.collection = collection.config.slug;
                user._strategy = 'local-jwt';
                done(null, user);
            } else {
                done(null, false);
            }
        } catch (err) {
            done(null, false);
        }
    });
};
