// payload/dist/admin/components/views/Version/RenderFieldsToDiff/fields/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _Iterable = /*#__PURE__*/ _interop_require_default(require("./Iterable"));
const _Nested = /*#__PURE__*/ _interop_require_default(require("./Nested"));
const _Relationship = /*#__PURE__*/ _interop_require_default(require("./Relationship"));
const _Select = /*#__PURE__*/ _interop_require_default(require("./Select"));
const _Tabs = /*#__PURE__*/ _interop_require_default(require("./Tabs"));
const _Text = /*#__PURE__*/ _interop_require_default(require("./Text"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const _default = {
    array: _Iterable.default,
    blocks: _Iterable.default,
    checkbox: _Text.default,
    code: _Text.default,
    collapsible: _Nested.default,
    date: _Text.default,
    email: _Text.default,
    group: _Nested.default,
    json: _Text.default,
    number: _Text.default,
    point: _Text.default,
    radio: _Select.default,
    relationship: _Relationship.default,
    richText: _Text.default,
    row: _Nested.default,
    select: _Select.default,
    tabs: _Tabs.default,
    text: _Text.default,
    textarea: _Text.default,
    upload: _Relationship.default
};
