// payload/dist/admin/components/views/Version/RenderFieldsToDiff/fields/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "diffStyles", {
    enumerable: true,
    get: function() {
        return diffStyles;
    }
});
const diffStyles = {
    variables: {
        dark: {
            addedBackground: 'var(--theme-success-900)',
            addedColor: 'var(--theme-success-100)',
            diffViewerBackground: 'transparent',
            diffViewerColor: 'var(--theme-text)',
            emptyLineBackground: 'var(--theme-elevation-50)',
            removedBackground: 'var(--theme-error-900)',
            removedColor: 'var(--theme-error-100)',
            wordAddedBackground: 'var(--theme-success-800)',
            wordRemovedBackground: 'var(--theme-error-800)'
        },
        light: {
            addedBackground: 'var(--theme-success-100)',
            addedColor: 'var(--theme-success-900)',
            diffViewerBackground: 'transparent',
            diffViewerColor: 'var(--theme-text)',
            emptyLineBackground: 'var(--theme-elevation-50)',
            removedBackground: 'var(--theme-error-100)',
            removedColor: 'var(--theme-error-900)',
            wordAddedBackground: 'var(--theme-success-200)',
            wordRemovedBackground: 'var(--theme-error-200)'
        }
    }
};
