// payload/dist/auth/graphql/resolvers/verifyEmail.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _isolateObjectProperty = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateObjectProperty"));
const _verifyEmail = /*#__PURE__*/ _interop_require_default(require("../../operations/verifyEmail"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function verifyEmailResolver(collection) {
    async function resolver(_, args, context) {
        let { req } = context;
        req = (0, _isolateObjectProperty.default)(req, 'locale');
        req = (0, _isolateObjectProperty.default)(req, 'fallbackLocale');
        if (args.locale) req.locale = args.locale;
        if (args.fallbackLocale) req.fallbackLocale = args.fallbackLocale;
        const options = {
            api: 'GraphQL',
            collection,
            req: (0, _isolateObjectProperty.default)(req, 'transactionID'),
            res: context.res,
            token: args.token
        };
        const success = await (0, _verifyEmail.default)(options);
        return success;
    }
    return resolver;
}
const _default = verifyEmailResolver;
