// payload/dist/admin/components/elements/FileDetails/Meta/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _formatFilesize = /*#__PURE__*/ _interop_require_default(require("../../../../../uploads/formatFilesize"));
const _Edit = /*#__PURE__*/ _interop_require_default(require("../../../icons/Edit"));
const _Config = require("../../../utilities/Config");
const _CopyToClipboard = /*#__PURE__*/ _interop_require_default(require("../../CopyToClipboard"));
const _DocumentDrawer = require("../../DocumentDrawer");
const _Tooltip = /*#__PURE__*/ _interop_require_default(require("../../Tooltip"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'file-meta';
const Meta = (props)=>{
    const { id, collection, filename, filesize, height, mimeType, staticURL, url, width } = props;
    const [hovered, setHovered] = (0, _react.useState)(false);
    const openInDrawer = Boolean(id && collection);
    const [DocumentDrawer, DocumentDrawerToggler] = (0, _DocumentDrawer.useDocumentDrawer)({
        id,
        collectionSlug: collection
    });
    const { serverURL } = (0, _Config.useConfig)();
    const fileURL = url || `${serverURL}${staticURL}/${filename}`;
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__url`
    }, openInDrawer && /*#__PURE__*/ _react.default.createElement(DocumentDrawer, null), /*#__PURE__*/ _react.default.createElement("a", {
        href: fileURL,
        rel: "noopener noreferrer",
        target: "_blank"
    }, filename), /*#__PURE__*/ _react.default.createElement(_CopyToClipboard.default, {
        defaultMessage: "Copy URL",
        value: fileURL
    }), openInDrawer && /*#__PURE__*/ _react.default.createElement(DocumentDrawerToggler, {
        className: `${baseClass}__edit`,
        onMouseEnter: ()=>setHovered(true),
        onMouseLeave: ()=>setHovered(false)
    }, /*#__PURE__*/ _react.default.createElement(_Edit.default, null), /*#__PURE__*/ _react.default.createElement(_Tooltip.default, {
        show: hovered
    }, "Edit"))), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__size-type`
    }, (0, _formatFilesize.default)(filesize), width && height && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, " - ", width, "x", height), mimeType && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, " - ", mimeType)));
};
const _default = Meta;
