// payload/dist/admin/components/elements/Header/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "AppHeader", {
    enumerable: true,
    get: function() {
        return AppHeader;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _Account = /*#__PURE__*/ _interop_require_default(require("../../graphics/Account"));
const _ActionsProvider = require("../../utilities/ActionsProvider");
const _Config = require("../../utilities/Config");
const _Hamburger = require("../Hamburger");
const _Localizer = /*#__PURE__*/ _interop_require_default(require("../Localizer"));
const _LocalizerLabel = require("../Localizer/LocalizerLabel");
const _NavToggler = require("../Nav/NavToggler");
const _context = require("../Nav/context");
const _StepNav = /*#__PURE__*/ _interop_require_default(require("../StepNav"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'app-header';
const AppHeader = ()=>{
    const { t } = (0, _reacti18next.useTranslation)();
    const { localization, routes: { admin: adminRoute } } = (0, _Config.useConfig)();
    const { actions } = (0, _ActionsProvider.useActions)();
    const { navOpen } = (0, _context.useNav)();
    const customControlsRef = (0, _react.useRef)(null);
    const [isScrollable, setIsScrollable] = (0, _react.useState)(false);
    (0, _react.useEffect)(()=>{
        const checkIsScrollable = ()=>{
            const el = customControlsRef.current;
            if (el) {
                const scrollable = el.scrollWidth > el.clientWidth;
                setIsScrollable(scrollable);
            }
        };
        checkIsScrollable();
        window.addEventListener('resize', checkIsScrollable);
        return ()=>{
            window.removeEventListener('resize', checkIsScrollable);
        };
    }, [
        actions
    ]);
    return /*#__PURE__*/ _react.default.createElement("header", {
        className: [
            baseClass,
            navOpen && `${baseClass}--nav-open`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bg`
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__content`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrapper`
    }, /*#__PURE__*/ _react.default.createElement(_NavToggler.NavToggler, {
        className: `${baseClass}__mobile-nav-toggler`,
        tabIndex: -1
    }, /*#__PURE__*/ _react.default.createElement(_Hamburger.Hamburger, null)), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__controls-wrapper`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__step-nav-wrapper`
    }, /*#__PURE__*/ _react.default.createElement(_StepNav.default, {
        className: `${baseClass}__step-nav`
    })), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__actions-wrapper`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__actions`,
        ref: customControlsRef
    }, Array.isArray(actions) && actions.map((Component, i)=>/*#__PURE__*/ _react.default.createElement("div", {
            className: isScrollable && i === actions.length - 1 ? `${baseClass}__last-action` : '',
            key: i
        }, /*#__PURE__*/ _react.default.createElement(Component, null)))), isScrollable && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__gradient-placeholder`
    })), localization && /*#__PURE__*/ _react.default.createElement(_LocalizerLabel.LocalizerLabel, {
        ariaLabel: "invisible",
        className: `${baseClass}__localizer-spacing`
    }), /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
        "aria-label": t('authentication:account'),
        className: `${baseClass}__account`,
        tabIndex: 0,
        to: `${adminRoute}/account`
    }, /*#__PURE__*/ _react.default.createElement(_Account.default, null))))), /*#__PURE__*/ _react.default.createElement(_Localizer.default, {
        className: `${baseClass}__localizer`
    }));
};
