// payload/dist/admin/components/views/Unauthorized/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _Button = /*#__PURE__*/ _interop_require_default(require("../../elements/Button"));
const _Minimal = /*#__PURE__*/ _interop_require_default(require("../../templates/Minimal"));
const _Auth = require("../../utilities/Auth");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const Unauthorized = ()=>{
    const { t } = (0, _reacti18next.useTranslation)('general');
    const config = (0, _Config.useConfig)();
    const { user } = (0, _Auth.useAuth)();
    const { admin: { logoutRoute }, routes: { admin } } = config;
    return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
        className: "unauthorized"
    }, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: t('error:unauthorized'),
        keywords: t('error:unauthorized'),
        title: t('error:unauthorized')
    }), /*#__PURE__*/ _react.default.createElement("h2", null, user ? t('general:unauthorized') : t('error:unauthorized')), /*#__PURE__*/ _react.default.createElement("p", null, t('error:notAllowedToAccessPage')), /*#__PURE__*/ _react.default.createElement("br", null), /*#__PURE__*/ _react.default.createElement(_Button.default, {
        el: "link",
        to: `${admin}${user ? '' : logoutRoute}`
    }, user ? t('general:backToDashboard') : t('authentication:logOut')));
};
const _default = Unauthorized;
