// payload/dist/admin/components/elements/CodeEditor/CodeEditor.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("@monaco-editor/react"));
const _react1 = /*#__PURE__*/ _interop_require_default(require("react"));
const _Theme = require("../../utilities/Theme");
const _ShimmerEffect = require("../ShimmerEffect");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'code-editor';
const CodeEditor = (props)=>{
    const { className, height, options, readOnly, ...rest } = props;
    const { theme } = (0, _Theme.useTheme)();
    const classes = [
        baseClass,
        className,
        rest?.defaultLanguage ? `language--${rest.defaultLanguage}` : ''
    ].filter(Boolean).join(' ');
    return /*#__PURE__*/ _react1.default.createElement(_react.default, {
        className: classes,
        height: height,
        loading: /*#__PURE__*/ _react1.default.createElement(_ShimmerEffect.ShimmerEffect, {
            height: height
        }),
        options: {
            detectIndentation: true,
            minimap: {
                enabled: false
            },
            readOnly: Boolean(readOnly),
            scrollBeyondLastLine: false,
            tabSize: 2,
            wordWrap: 'on',
            ...options
        },
        theme: theme === 'dark' ? 'vs-dark' : 'vs',
        ...rest
    });
};
const _default = CodeEditor;
