// payload/dist/admin/api.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "requests", {
    enumerable: true,
    get: function() {
        return requests;
    }
});
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const requests = {
    delete: (url, options = {
        headers: {}
    })=>{
        const headers = options && options.headers ? {
            ...options.headers
        } : {};
        const formattedOptions = {
            ...options,
            credentials: 'include',
            headers: {
                ...headers
            },
            method: 'delete'
        };
        return fetch(url, formattedOptions);
    },
    get: (url, options = {
        headers: {}
    })=>{
        let query = '';
        if (options.params) {
            query = _qs.default.stringify(options.params, {
                addQueryPrefix: true
            });
        }
        return fetch(`${url}${query}`, {
            credentials: 'include',
            ...options
        });
    },
    patch: (url, options = {
        headers: {}
    })=>{
        const headers = options && options.headers ? {
            ...options.headers
        } : {};
        const formattedOptions = {
            ...options,
            credentials: 'include',
            headers: {
                ...headers
            },
            method: 'PATCH'
        };
        return fetch(url, formattedOptions);
    },
    post: (url, options = {
        headers: {}
    })=>{
        const headers = options && options.headers ? {
            ...options.headers
        } : {};
        const formattedOptions = {
            ...options,
            credentials: 'include',
            headers: {
                ...headers
            },
            method: 'post'
        };
        return fetch(`${url}`, formattedOptions);
    },
    put: (url, options = {
        headers: {}
    })=>{
        const headers = options && options.headers ? {
            ...options.headers
        } : {};
        const formattedOptions = {
            ...options,
            credentials: 'include',
            headers: {
                ...headers
            },
            method: 'put'
        };
        return fetch(url, formattedOptions);
    }
};
