// payload/dist/admin/components.config.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _cssminimizerwebpackplugin = /*#__PURE__*/ _interop_require_default(require("css-minimizer-webpack-plugin"));
const _minicssextractplugin = /*#__PURE__*/ _interop_require_default(require("mini-css-extract-plugin"));
const _path = /*#__PURE__*/ _interop_require_default(require("path"));
const _terserwebpackplugin = /*#__PURE__*/ _interop_require_default(require("terser-webpack-plugin"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const componentWebpackConfig = {
    entry: {
        main: [
            _path.default.resolve(__dirname, './components/index.js')
        ]
    },
    externals: {
        react: 'react'
    },
    mode: 'production',
    module: {
        rules: [
            {
                exclude: /node_modules/,
                test: /\.(t|j)sx?$/,
                use: [
                    {
                        loader: require.resolve('swc-loader')
                    }
                ]
            },
            {
                oneOf: [
                    {
                        loader: require.resolve('url-loader'),
                        options: {
                            name: 'static/media/[name].[hash:8].[ext]',
                            limit: 10000
                        },
                        test: [
                            /\.bmp$/,
                            /\.gif$/,
                            /\.jpe?g$/,
                            /\.png$/
                        ]
                    },
                    {
                        sideEffects: true,
                        test: /\.(sa|sc|c)ss$/,
                        use: [
                            _minicssextractplugin.default.loader,
                            'css-loader',
                            {
                                loader: 'postcss-loader',
                                options: {
                                    postcssOptions: {
                                        plugins: [
                                            'postcss-preset-env'
                                        ]
                                    }
                                }
                            },
                            'sass-loader'
                        ]
                    },
                    {
                        exclude: [
                            /\.(js|jsx|mjs)$/,
                            /\.html$/,
                            /\.json$/
                        ],
                        loader: require.resolve('file-loader'),
                        options: {
                            name: 'static/media/[name].[hash:8].[ext]'
                        }
                    }
                ]
            }
        ]
    },
    optimization: {
        minimizer: [
            new _terserwebpackplugin.default({
                extractComments: false
            }),
            new _cssminimizerwebpackplugin.default({})
        ]
    },
    output: {
        filename: 'index.js',
        libraryTarget: 'commonjs2',
        path: _path.default.resolve(__dirname, '../../components'),
        publicPath: '/'
    },
    plugins: [
        new _minicssextractplugin.default({
            filename: 'styles.css',
            ignoreOrder: true
        })
    ],
    resolve: {
        alias: {
            'payload-scss-overrides': _path.default.resolve(__dirname, './scss/overrides.scss')
        },
        modules: [
            'node_modules',
            _path.default.resolve(__dirname, '../../node_modules')
        ]
    },
    stats: 'errors-only'
};
const _default = componentWebpackConfig;
