// payload/dist/admin/Root.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _modal = require("@faceless-ui/modal");
const _scrollinfo = require("@faceless-ui/scroll-info");
const _windowinfo = require("@faceless-ui/window-info");
const _payloadconfig = /*#__PURE__*/ _interop_require_default(require("payload-config"));
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reactrouterdom = require("react-router-dom");
const _reacttoastify = require("react-toastify");
const _context = require("./components/elements/Nav/context");
const _StepNav = require("./components/elements/StepNav");
const _Auth = require("./components/utilities/Auth");
const _Config = require("./components/utilities/Config");
const _CustomProvider = require("./components/utilities/CustomProvider");
const _DocumentEvents = require("./components/utilities/DocumentEvents");
const _I18n = require("./components/utilities/I18n");
const _LanguageWrap = require("./components/utilities/LanguageWrap");
const _LoadingOverlay = require("./components/utilities/LoadingOverlay");
const _Locale = require("./components/utilities/Locale");
const _Preferences = require("./components/utilities/Preferences");
const _SearchParams = require("./components/utilities/SearchParams");
const _Theme = require("./components/utilities/Theme");
const _Routes = require("./components/views/Routes");
require("./scss/app.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const Root = ({ config: incomingConfig })=>{
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Config.ConfigProvider, {
        config: incomingConfig || _payloadconfig.default
    }, /*#__PURE__*/ _react.default.createElement(_I18n.I18n, null), /*#__PURE__*/ _react.default.createElement(_windowinfo.WindowInfoProvider, {
        breakpoints: {
            l: '(max-width: 1440px)',
            m: '(max-width: 1024px)',
            s: '(max-width: 768px)',
            xs: '(max-width: 400px)'
        }
    }, /*#__PURE__*/ _react.default.createElement(_scrollinfo.ScrollInfoProvider, null, /*#__PURE__*/ _react.default.createElement(_reactrouterdom.BrowserRouter, null, /*#__PURE__*/ _react.default.createElement(_modal.ModalProvider, {
        classPrefix: "payload",
        transTime: 0,
        zIndex: "var(--z-modal)"
    }, /*#__PURE__*/ _react.default.createElement(_Auth.AuthProvider, null, /*#__PURE__*/ _react.default.createElement(_Preferences.PreferencesProvider, null, /*#__PURE__*/ _react.default.createElement(_Theme.ThemeProvider, null, /*#__PURE__*/ _react.default.createElement(_LanguageWrap.LanguageWrap, null, /*#__PURE__*/ _react.default.createElement(_SearchParams.SearchParamsProvider, null, /*#__PURE__*/ _react.default.createElement(_Locale.LocaleProvider, null, /*#__PURE__*/ _react.default.createElement(_StepNav.StepNavProvider, null, /*#__PURE__*/ _react.default.createElement(_LoadingOverlay.LoadingOverlayProvider, null, /*#__PURE__*/ _react.default.createElement(_DocumentEvents.DocumentEventsProvider, null, /*#__PURE__*/ _react.default.createElement(_context.NavProvider, null, /*#__PURE__*/ _react.default.createElement(_CustomProvider.CustomProvider, null, /*#__PURE__*/ _react.default.createElement(_Routes.Routes, null)))))))))), /*#__PURE__*/ _react.default.createElement(_modal.ModalContainer, null)))))))), /*#__PURE__*/ _react.default.createElement(_reacttoastify.ToastContainer, {
        icon: false,
        position: "bottom-center",
        transition: _reacttoastify.Slide
    }));
};
const _default = Root;
