// payload/dist/admin/components/elements/Localizer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _qs = /*#__PURE__*/ _interop_require_default(require("qs"));
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../utilities/getTranslation");
const _Config = require("../../utilities/Config");
const _Locale = require("../../utilities/Locale");
const _SearchParams = require("../../utilities/SearchParams");
const _Popup = /*#__PURE__*/ _interop_require_default(require("../Popup"));
const _PopupButtonList = /*#__PURE__*/ _interop_require_wildcard(require("../Popup/PopupButtonList"));
const _LocalizerLabel = require("./LocalizerLabel");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'localizer';
const Localizer = (props)=>{
    const { className } = props;
    const config = (0, _Config.useConfig)();
    const { localization } = config;
    const { i18n } = (0, _reacti18next.useTranslation)();
    const locale = (0, _Locale.useLocale)();
    const searchParams = (0, _SearchParams.useSearchParams)();
    const localeLabel = (0, _getTranslation.getTranslation)(locale.label, i18n);
    if (localization) {
        const { locales } = localization;
        return /*#__PURE__*/ _react.default.createElement("div", {
            className: [
                baseClass,
                className
            ].filter(Boolean).join(' ')
        }, /*#__PURE__*/ _react.default.createElement(_Popup.default, {
            button: /*#__PURE__*/ _react.default.createElement(_LocalizerLabel.LocalizerLabel, null),
            horizontalAlign: "right",
            render: ({ close })=>/*#__PURE__*/ _react.default.createElement(_PopupButtonList.ButtonGroup, null, /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, locale ? /*#__PURE__*/ _react.default.createElement(_PopupButtonList.Button, {
                    active: true,
                    key: locale.code,
                    onClick: close,
                    to: {
                        search: _qs.default.stringify({
                            ...searchParams,
                            locale: locale.code
                        })
                    }
                }, localeLabel, localeLabel !== locale.code && ` (${locale.code})`) : null, locales.map((localeOption)=>{
                    if (locale.code === localeOption.code) return null;
                    const newParams = {
                        ...searchParams,
                        locale: localeOption.code
                    };
                    const search = _qs.default.stringify(newParams);
                    const localeOptionLabel = (0, _getTranslation.getTranslation)(localeOption.label, i18n);
                    return /*#__PURE__*/ _react.default.createElement(_PopupButtonList.Button, {
                        key: localeOption.code,
                        onClick: close,
                        to: {
                            search
                        }
                    }, localeOptionLabel, localeOptionLabel !== localeOption.code && ` (${localeOption.code})`);
                }))),
            showScrollbar: true,
            size: "large"
        }));
    }
    return null;
};
const _default = Localizer;
