// payload/dist/admin/components/elements/GenerateConfirmation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _modal = require("@faceless-ui/modal");
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _reacttoastify = require("react-toastify");
const _Minimal = /*#__PURE__*/ _interop_require_default(require("../../templates/Minimal"));
const _DocumentInfo = require("../../utilities/DocumentInfo");
const _Button = /*#__PURE__*/ _interop_require_default(require("../Button"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'generate-confirmation';
const GenerateConfirmation = (props)=>{
    const { highlightField, setKey } = props;
    const { id } = (0, _DocumentInfo.useDocumentInfo)();
    const { toggleModal } = (0, _modal.useModal)();
    const { t } = (0, _reacti18next.useTranslation)('authentication');
    const modalSlug = `generate-confirmation-${id}`;
    const handleGenerate = ()=>{
        setKey();
        toggleModal(modalSlug);
        _reacttoastify.toast.success(t('newAPIKeyGenerated'), {
            autoClose: 3000
        });
        highlightField(true);
    };
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Button.default, {
        buttonStyle: "secondary",
        onClick: ()=>{
            toggleModal(modalSlug);
        },
        size: "small"
    }, t('generateNewAPIKey')), /*#__PURE__*/ _react.default.createElement(_modal.Modal, {
        className: baseClass,
        slug: modalSlug
    }, /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
        className: `${baseClass}__template`
    }, /*#__PURE__*/ _react.default.createElement("h1", null, t('confirmGeneration')), /*#__PURE__*/ _react.default.createElement("p", null, /*#__PURE__*/ _react.default.createElement(_reacti18next.Trans, {
        i18nKey: "generatingNewAPIKeyWillInvalidate",
        t: t
    }, "generatingNewAPIKeyWillInvalidate", /*#__PURE__*/ _react.default.createElement("strong", null, "invalidate"))), /*#__PURE__*/ _react.default.createElement(_Button.default, {
        buttonStyle: "secondary",
        onClick: ()=>{
            toggleModal(modalSlug);
        },
        type: "button"
    }, t('general:cancel')), /*#__PURE__*/ _react.default.createElement(_Button.default, {
        onClick: handleGenerate
    }, t('generate')))));
};
const _default = GenerateConfirmation;
